{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homology function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "imtesting = True # After done testing, remove every condition this appears"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# TODO: Will we ever need to change this configuration parameters?\n",
    "# TODO: Set it to use the GPU and all CPU threads? yasara.Processors(cputhreads=None, gpu=None)\n",
    "\n",
    "import yasara as y\n",
    "y.info.mode='txt'\n",
    "def yasara_homology(filename,            in_psiblasts=3,          in_evalue=0.5, \\\n",
    "                    in_templates=5,      in_alignments=5,         in_oligostate=4,\\\n",
    "                    in_termextension=10, in_structprofile='try'):\n",
    "    \"\"\"\"\"\"\n",
    "    print(\"Homology experiment:\", filename)\n",
    "    y.ExperimentHomologyModeling(sequencefile=filename,\\\n",
    "                                psiblasts=in_psiblasts,\\\n",
    "                                evalue=in_evalue, \\\n",
    "                                templates=in_templates, \\\n",
    "                                alignments=in_alignments,\\\n",
    "                                oligostate=in_oligostate,\\\n",
    "                                termextension=in_termextension,\\\n",
    "                                structprofile=in_structprofile)\n",
    "    y.Experiment(\"On\")\n",
    "    y.Wait(\"ExpEnd\")\n",
    "    print(\"Exp ended\")\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to Analyze the Quality of the homologies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('-0.908', 'IL-18(189).yob')\n",
      "('-2.388', 'IL-18(193)_4xfs-a.yob')\n",
      "('-0.956', 'IL-1A(271).yob')\n",
      "('-1.199', 'IL-1B(268).yob')\n",
      "('-4.434', 'IL-1B(300)_1twm-a.yob')\n",
      "('-1.442', 'IL-1Ra(143)_1ilr-~.yob')\n",
      "('-1.054', 'IL-1Ra(159).yob')\n",
      "('-0.820', 'IL-1Ra(177).yob')\n",
      "('-1.138', 'IL-1Ra(180).yob')\n",
      "('-1.305', 'IL-33(144).yob')\n",
      "('-1.703', 'IL-33(228).yob')\n",
      "('-1.334', 'IL-33(270).yob')\n",
      "('-1.487', 'IL-36Ra(155)_1md6-a01.yob')\n",
      "('-2.014', 'IL-37(157)_5hn1-~02.yob')\n",
      "('-2.203', 'IL-37(178)_5hn1-~05.yob')\n",
      "('-2.178', 'IL-37(192)_5hn1-~01.yob')\n",
      "('-2.353', 'IL-37(197)_5hn1-~01.yob')\n",
      "('-2.910', 'IL-37(218)_5hn1-~01.yob')\n",
      "('-1.020', 'IL-38(152).yob')\n",
      "('0.978', 'IL-6(212).yob')\n",
      "('-0.144', 'IL-8(99).yob')\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "h_one = re.compile(\"This hybrid model with Z-score (.{,10}) was saved as the final one, <.*?>(.*\\.yob)\")\n",
    "\n",
    "single_z = re.compile(\"The model with an overall quality Z-score of (.{,10}) has been saved as \")\n",
    "single_yob = re.compile(\"Instead, the model .*? was simply saved as the final one <.*?>(.*?yob).\")\n",
    "\n",
    "no_hyb = re.compile(\"The hybrid model was discarded, and (.*?yob) was saved as the final model\")\n",
    "no_hyb_table = re.compile(\"<table><tr><td>Transfer.*?</td></tr></table>\")\n",
    "no_hyb_z = re.compile(\"<b>(.*?)</b>\")\n",
    "\n",
    "\n",
    "def homology_qa(folder):\n",
    "    \"\"\"Obtain z-scores and the final yasara object for a simulation.\n",
    "    \n",
    "    \"\"\"\n",
    "    global imtesting\n",
    "    if imtesting:\n",
    "        folder = \"Saved_run/\"+folder\n",
    "    with open(folder+'/'+folder+'.html') as f:\n",
    "        html = f.read() # load html as text\n",
    "    html = html[html.find(\"The model ranking\"):] # discard the initil analysis, keep the summary\n",
    "    html = \"\".join(html.splitlines()) # remove newline symbols\n",
    "    \n",
    "    # Try to obtain results for single hybrid model\n",
    "    re_obj = h_one.search(html)\n",
    "    if re_obj:\n",
    "        return re_obj.groups()\n",
    "    re_obj = single_z.search(html)\n",
    "    # Try to obtain results for single model\n",
    "    if re_obj:\n",
    "        re_obj1 = single_yob.search(html)\n",
    "        if re_obj1:\n",
    "            return (re_obj.group(1), re_obj1.group(1))\n",
    "    re_obj = no_hyb.search(html)\n",
    "    # Try to obtain results for a multi model tthat couldn't be hybridized\n",
    "    if re_obj:\n",
    "        table = no_hyb_table.search(html)\n",
    "        if table:\n",
    "            re_obj1 = no_hyb_z.search(table.group(0))\n",
    "            if re_obj1:\n",
    "                return (re_obj1.group(1), re_obj.group(1))\n",
    "    # if nothing was found, warn the user # TODO: or the system admin\n",
    "    print(\"Error: Could not parse file '\"+folder+\"' properly\")\n",
    "    return False\n",
    "\n",
    "if True:\n",
    "    print (homology_qa(\"IL-18(189)\"))\n",
    "    print (homology_qa(\"IL-18(193)\"))\n",
    "    print (homology_qa(\"IL-1A(271)\"))\n",
    "    print (homology_qa(\"IL-1B(268)\"))\n",
    "    print (homology_qa(\"IL-1B(300)\"))\n",
    "    print (homology_qa(\"IL-1Ra(143)\"))\n",
    "    print (homology_qa(\"IL-1Ra(159)\"))\n",
    "    print (homology_qa(\"IL-1Ra(177)\"))\n",
    "    print (homology_qa(\"IL-1Ra(180)\"))\n",
    "    print (homology_qa(\"IL-33(144)\"))\n",
    "    print (homology_qa(\"IL-33(228)\"))\n",
    "    print (homology_qa(\"IL-33(270)\"))\n",
    "    print (homology_qa(\"IL-36Ra(155)\"))\n",
    "    print (homology_qa(\"IL-37(157)\"))\n",
    "    print (homology_qa(\"IL-37(178)\"))\n",
    "    print (homology_qa(\"IL-37(192)\"))\n",
    "    print (homology_qa(\"IL-37(197)\"))\n",
    "    print (homology_qa(\"IL-37(218)\"))\n",
    "    print (homology_qa(\"IL-38(152)\"))\n",
    "    print (homology_qa(\"IL-6(212)\"))\n",
    "    print (homology_qa(\"IL-8(99)\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IL-1A(271)/IL-1A(271).fasta\n",
      "Exp ended\n",
      "IL-1B(268)/IL-1B(268).fasta\n",
      "Exp ended\n",
      "IL-1Ra(177)/IL-1Ra(177).fasta\n",
      "Exp ended\n",
      "IL-1Ra(159)/IL-1Ra(159).fasta\n",
      "Exp ended\n",
      "IL-1Ra(180)/IL-1Ra(180).fasta\n",
      "Exp ended\n",
      "IL-1Ra(143)/IL-1Ra(143).fasta\n",
      "Exp ended\n",
      "IL-18(193)/IL-18(193).fasta\n",
      "Exp ended\n",
      "IL-18(189)/IL-18(189).fasta\n",
      "Exp ended\n",
      "IL-33(270)/IL-33(270).fasta\n",
      "Exp ended\n",
      "IL-33(228)/IL-33(228).fasta\n",
      "Exp ended\n",
      "IL-33(144)/IL-33(144).fasta\n",
      "Exp ended\n",
      "IL-36Ra(155)/IL-36Ra(155).fasta\n",
      "Exp ended\n",
      "IL-37(218)/IL-37(218).fasta\n",
      "Exp ended\n",
      "IL-37(192)/IL-37(192).fasta\n",
      "Exp ended\n",
      "IL-37(178)/IL-37(178).fasta\n",
      "Exp ended\n",
      "IL-37(197)/IL-37(197).fasta\n",
      "Exp ended\n",
      "IL-37(157)/IL-37(157).fasta\n",
      "Exp ended\n",
      "IL-38(152)/IL-38(152).fasta\n",
      "Exp ended\n",
      "IL-6(212)/IL-6(212).fasta\n",
      "Exp ended\n",
      "IL-8(99)/IL-8(99).fasta\n",
      "Exp ended\n"
     ]
    }
   ],
   "source": [
    "seq_dict = fasta_to_seq(\"IL-1_formatted.fastaR\")\n",
    "organize_sequences(seq_dict)\n",
    "for folder in seq_dict.keys():\n",
    "    print(folder)\n",
    "    yasara_homology(folder+'/'+folder+'.fasta')\n",
    "    z_score, yas_obj = homology_qa(folder) # TODO: Store and/or print the z-score\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using a function to run the Dock Ensembl macro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Docking experiment:Saved_run/IL-36Ra(155)\n",
      "Exp Saved_run/IL-36Ra(155) Ended\n",
      "Docking experiment:Saved_run/IL-1Ra(143)\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Connection to YASARA broken. Either you exited YASARA manually or it encountered a fatal error",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionResetError\u001b[0m                      Traceback (most recent call last)",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mreceive\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    731\u001b[0m       \u001b[0;31m# BUT RAISES AN EXCEPTION (WHICH ONE??) IN SOME WINDOWS VERSIONS\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 732\u001b[0;31m       \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mchunk\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    733\u001b[0m       \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mConnectionResetError\u001b[0m: [Errno 104] Connection reset by peer",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-c85cd4e97237>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[0myasara_docking\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"IL-36Ra(155)\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"IL-36Ra(155)_1md6-a01.yob\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m \u001b[0myasara_docking\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"IL-1Ra(143)\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"IL-1Ra(143).yob\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-c85cd4e97237>\u001b[0m in \u001b[0;36myasara_docking\u001b[0;34m(folder, receptor_file, method)\u001b[0m\n\u001b[1;32m     50\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Error: Receptor file not found for \"\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0mfolder\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# TODO: List of warnings and errors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Docking experiment:\"\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0mfolder\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m     \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mApplyMacro\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"dock_runensemble_dummy.mcr\"\u001b[0m\u001b[0;34m,\u001b[0m                  \u001b[0mtargets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnewrec\u001b[0m\u001b[0;34m,\u001b[0m                  \u001b[0mremove\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"FromUnderscore\"\u001b[0m\u001b[0;34m,\u001b[0m                  \u001b[0mnewextension\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Exp\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mfolder\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Ended\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[0;31m# y.Exit()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mApplyMacro\u001b[0;34m(filename, targets, remove, newextension)\u001b[0m\n\u001b[1;32m   1608\u001b[0m   \u001b[0mcommand\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;34m'Remove='\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mcstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1609\u001b[0m   \u001b[0mcommand\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;34m'NewExtension='\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mcstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnewextension\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1610\u001b[0;31m   \u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1611\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1612\u001b[0m \u001b[0;31m# SWITCH PLASMA INSIDE ATOMS ON/OFF\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(command)\u001b[0m\n\u001b[1;32m   1023\u001b[0m \u001b[0;31m# =======================\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1024\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1025\u001b[0;31m   \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\n\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocess\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"EXECUTE\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1026\u001b[0m   \u001b[0;31m# THIS IS A MULTI-LINE COMMAND, SEND EACH LINE SEPARATELY SO THAT ALL ERRORS CAN BE CAUGHT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1027\u001b[0m   \u001b[0mcomlist\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\n\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mprocess\u001b[0;34m(messagetype, messagedata)\u001b[0m\n\u001b[1;32m   1017\u001b[0m   \u001b[0;31m# DO NOT CATCH CONNECTION PROBLEMS, HANDLE THE EXCEPTION NORMALLY\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1018\u001b[0m   \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msendmessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessagetype\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmessagedata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1019\u001b[0;31m   \u001b[0mresult\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreceivemessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRESULT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1020\u001b[0m   \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1021\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mreceivemessage\u001b[0;34m(self, expectedtype)\u001b[0m\n\u001b[1;32m    750\u001b[0m   \u001b[0;31m# =============================\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    751\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0mreceivemessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mexpectedtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 752\u001b[0;31m     \u001b[0;34m(\u001b[0m\u001b[0mmessagetype\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdatasize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstruct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munpack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ii'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreceive\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    753\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreceive\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatasize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    754\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmessagetype\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mATOMDATA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mreceive\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    732\u001b[0m       \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mchunk\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    733\u001b[0m       \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 734\u001b[0;31m       \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreportbrokenconn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    735\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchunk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreportbrokenconn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    736\u001b[0m       \u001b[0mdata\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0mchunk\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Downloads/yasara/pym/yasaramodule.py\u001b[0m in \u001b[0;36mreportbrokenconn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    704\u001b[0m     \u001b[0mpid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 705\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Connection to YASARA broken. Either you exited YASARA manually or it encountered a fatal error\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    706\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    707\u001b[0m   \u001b[0;31m# SEND DATA TO YASARA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Connection to YASARA broken. Either you exited YASARA manually or it encountered a fatal error"
     ]
    }
   ],
   "source": [
    "imtesting = True # After done testing, remove every condition this appears (keep any 'else' clause)\n",
    "import yasara as y\n",
    "\n",
    "y.info.mode='txt' \n",
    "\n",
    "# TODO: Adjust the parameters inside the file\n",
    "# y.CellAuto(extension=17, shape=\"Cuboid\", selection1=\"Obj 1 Res Pro 41\")\n",
    "# Cell Auto,extension=17, Shape=Cuboid, Res Tyr 29\n",
    "\n",
    "import shutil\n",
    "from os import rename\n",
    "from os.path import isfile\n",
    "\n",
    "def parametrize_macro(flexres=\"flexres = ''\", celldef = \"\"):\n",
    "    \"\"\" Creates a parametrized macro from the model.\"\"\"\n",
    "    macromodel = \"dock_model.mcr\"\n",
    "    macro = \"dock_runensemble_dummy.mcr\"\n",
    "\n",
    "    with open(macromodel) as f:\n",
    "        macrotext = f.read()\n",
    "    macrotext = flexres.join(macrotext.split(\"# $$FLEXRES$$\"))\n",
    "    macrotext = celldef.join(macrotext.split(\"# $$CELL$$\"))\n",
    "    with open(macro, \"w\") as f:\n",
    "        f.write(macrotext)\n",
    "\n",
    "\n",
    "def yasara_docking(folder, receptor_file, method='VINA'):\n",
    "    global imtesting\n",
    "    if imtesting:\n",
    "        folder = \"Saved_run/\"+folder\n",
    "    filename = \".\".join(receptor_file.split(\".\")[:-1])\n",
    "    fileext = receptor_file.split(\".\")[-1]\n",
    "    ligmodel = \"a_ligand.pdb\"\n",
    "    newlig = folder + \"/\"+ filename + \"_ligand.pdb\"\n",
    "    \n",
    "    \n",
    "    parametrize_macro(celldef = \"Cell Auto,extension=17, Shape=Cuboid, Res Tyr 29\")\n",
    "    \n",
    "    # copies the ligand to the receptor's folder with the proper name\n",
    "    if not isfile(newlig):\n",
    "        shutil.copyfile(ligmodel, newlig)\n",
    "\n",
    "    # renames the receptor's yasara object adding _receptor\n",
    "    oldrec = folder + \"/\" + receptor_file\n",
    "    newrec = folder + \"/\" + filename+ \"_receptor.\" + fileext\n",
    "    if isfile(oldrec):\n",
    "        rename(oldrec, newrec)\n",
    "    else:\n",
    "        if not isfile(newrec):\n",
    "            print(\"Error: Receptor file not found for \"+ folder) # TODO: List of warnings and errors\n",
    "    print(\"Docking experiment:\"+ folder )\n",
    "    y.ApplyMacro(filename=\"dock_runensemble_dummy.mcr\", \\\n",
    "                 targets=newrec, \\\n",
    "                 remove=\"FromUnderscore\", \\\n",
    "                 newextension=\"\")\n",
    "    print(\"Exp\",folder,\"Ended\")\n",
    "    \n",
    "yasara_docking(\"IL-36Ra(155)\", \"IL-36Ra(155)_1md6-a01.yob\")\n",
    "yasara_docking(\"IL-1Ra(143)\", \"IL-1Ra(143).yob\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'if __name__ == \\'__main__\\':\\n    # assert len(sys.argv) == 2, \\'Usage is \"python organize_seqs <fastafile>\"\\'\\n    seq_dict = fasta_to_seq(sys.argv[1])\\n    organize_sequences(seq_dict)\\n    for filename in seq_dict.keys():\\n        yasara_homology(filename)'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"if __name__ == '__main__':\n",
    "    # assert len(sys.argv) == 2, 'Usage is \"python organize_seqs <fastafile>\"'\n",
    "    seq_dict = fasta_to_seq(sys.argv[1])\n",
    "    organize_sequences(seq_dict)\n",
    "    for filename in seq_dict.keys():\n",
    "        yasara_homology(filename)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selection cell "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yasara as y\n",
    "y.info.mode=\"txt\"\n",
    "y.ApplyMacro(filename=\"/home/imhof_team/Public/mauricio/workflow/dummy.mcr\", \\\n",
    "            targets=\"Saved_run/IL-36Ra(155)/IL-36Ra(155)_1md6-a01_receptor.yob\", \\\n",
    "            remove=\"FromUnderscore\", \\\n",
    "            newextension=\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# YASARA MACRO\n",
      "# TOPIC:       5. Structure prediction\n",
      "# TITLE:       Docking a ligand to a receptor ensemble with flexible side-chains \n",
      "# REQUIRES:    Structure\n",
      "# AUTHOR:      Elmar Krieger\n",
      "# LICENSE:     GPL\n",
      "# DESCRIPTION: This macro predicts the structure of a ligand-receptor complex. Receptor flexibility is considered by creating a receptor ensemble with alternative high-scoring solutions of the side-chain rotamer network. An analysis log file is written at the end. The macro can also continue a docking run that got interrupted, especially if the scene has not been moved or rotated manually during the docking. \n",
      "\n",
      "# Parameter section - adjust as needed, but NOTE that some changes only take effect\n",
      "# if you start an entirely new docking job, not if you continue an existing one. \n",
      "# =================================================================================\n",
      "\n",
      "# You can either set the target structure by clicking on Options > Macro > Set target,\n",
      "# by providing it as command line argument (see docs at Essentials > The command line),\n",
      "# or by uncommenting the line below and specifying it directly.\n",
      "#MacroTarget='/home/imhof_team/Public/mauricio/IL-1_homology/a'\n",
      "\n",
      "# Docking method, either AutoDockLGA or VINA\n",
      "method='VINA'\n",
      "\n",
      "# Size of the receptor ensemble with alternative side-chain rotamers\n",
      "receptors=5\n",
      "\n",
      "# Number of docking runs per receptor ensemble member, '0' leaves the choice to YASARA.\n",
      "# (maximally 999, each run can take up to an hour).\n",
      "# The total number of docking runs is receptors*receptorruns.\n",
      "receptorruns=0\n",
      "\n",
      "# Docking results usually cluster around certain hot spot conformations,\n",
      "# and the lowest energy complex in each cluster is saved. Two complexes belong to\n",
      "# different clusters if the ligand RMSD is larger than this minimum [A]:\n",
      "rmsdmin=5.0\n",
      "\n",
      "# Set to 0 if you don't want to merge clusters from the receptor ensemble members\n",
      "# in the end (merging keeps only one cluster within the 'rmsdmin' limit set above).\n",
      "merged=1\n",
      "\n",
      "# Set to 1 to keep the ligand completely rigid (alternatively you can provide\n",
      "# the ligand as a *.yob file and fix certain dihedral angles only).\n",
      "rigid=0\n",
      "\n",
      "# A selection of receptor residues to keep flexible, e.g. flexres = 'Lys 91 Leu 100',\n",
      "# or (if the receptor is not a monomer) flexres = 'Res Lys 91 Mol A or Res Leu 100 Mol B'.\n",
      "# Alternatively you can provide the receptor as a *.sce or *.yob file with fixed\n",
      "# atoms, which gives better control (e.g. you can keep only part of a side-chain\n",
      "# or even a terminal backbone flexible). \n",
      "flexres='' \n",
      "flexres = ''\n",
      "\n",
      "# Force field used for charge assignment, not used by VINA.\n",
      "ForceField AMBER03\n",
      "\n",
      "# The docking temperature. This currently affects the receptor flexibility only\n",
      "temp='298K' \n",
      "\n",
      "# Flag if a scene with all results should be saved for visualization with dock_play.mcr.\n",
      "# If you use a large receptor ensemble with large ligands and many docking runs, it may\n",
      "# exhaust the available memory to show them all together on screen. In this case set\n",
      "# scesaved to 0. The docking result player will then only show clusters but not all results,\n",
      "# which can of course still be found in the logfile.\n",
      "scesaved=1 \n",
      "\n",
      "\n",
      "# Normally no change required below this point\n",
      "# ============================================\n",
      "\n",
      "chargefof = ForceField\n",
      "if !receptorruns\n",
      "  # For an optimum speed/accuracy tradeoff, we select at least 20 runs, but more if CPUs are available\n",
      "  cpus = Processors\n",
      "  receptorruns=cpus\n",
      "  while receptorruns<20\n",
      "    receptorruns=receptorruns+cpus\n",
      "\n",
      "# Sanity checks\n",
      "if MacroTarget==''\n",
      "  RaiseError \"This macro requires a target. Either edit the macro file or click Options > Macro > Set target to choose a target structure\"\n",
      "if receptors>999 or receptorruns>999 or receptors*receptorruns>9999\n",
      "  RaiseError 'Too many docking runs selected, (receptors)/(receptorruns) would take forever'\n",
      "\n",
      "structlist='receptor','ligand'\n",
      "\n",
      "Clear\n",
      "# Do we have a user-supplied ensemble?\n",
      "ensfilename='(MacroTarget)_receptor_ensemble.sce'\n",
      "scene = FileSize (ensfilename)\n",
      "if !scene\n",
      "  # Do we already have an automatically generated receptor ensemble scene?\n",
      "  ensfilename='(MacroTarget)_receptor_ensemble(receptors).sce'\n",
      "scene = FileSize (ensfilename)\n",
      "if !scene\n",
      "  # Not yet. Do we have a receptor scene?\n",
      "  scefilename='(MacroTarget)_receptor.sce'\n",
      "  scene = FileSize (scefilename)\n",
      "  if !scene\n",
      "    # No, load PDB or YOB file of receptor and ligand\n",
      "    for struct in structlist\n",
      "      (struct)=0\n",
      "      for type in 'yob','pdb','sdf'\n",
      "        filename='(MacroTarget)_(struct).(type)'\n",
      "        exists = FileSize (filename)\n",
      "        if exists and (type!='sdf' or struct=='ligand')\n",
      "          (struct) = Load(type) (filename)\n",
      "          break\n",
      "      if not (struct)\n",
      "        RaiseError 'The (struct) was not found at (filename)'\n",
      "    # Orient the receptor, create a docking cell that includes the entire receptor \n",
      "    # and also has enough empty space around to accomodate the ligand (which has to\n",
      "    # be removed temporarily, since 'Cell Auto' encloses the entire soup). \n",
      "    ligandradius = RadiusObj (ligand)  \n",
      "    NiceOriObj (receptor)\n",
      "    # Ligand was only needed to determine the required cell size, will be loaded later\n",
      "    DelObj (ligand)\n",
      "    Cell Auto,Extension=(ligandradius*2)\n",
      "  else\n",
      "    # Load receptor scene\n",
      "    LoadSce (scefilename)\n",
      "    # Verify that the cell is present\n",
      "    simcell = CountObj SimCell\n",
      "    if !simcell\n",
      "      RaiseError 'If you provide a scene, it must contain a simulation cell, but none was found in (scefilename)'\n",
      "    if Objects>2\n",
      "      # We can't continue, because the ensemble can only be created for a single object\n",
      "      RaiseError 'Your scene contains (Objects) objects, while only the receptor and the docking cell are expected.'\n",
      "  # Warn if the cell is too large\n",
      "  x,y,z = Cell\n",
      "  if x>96 or y>96 or z>96\n",
      "    ShowMessage \"A cell axis is longer than 96 A, which may reduce docking accuracy. Consider focusing on the active site.\"\n",
      "    Wait ContinueButton\n",
      "    HideMessage\n",
      "  # To create a receptor ensemble, the entire receptor must be inside the cell,\n",
      "  # so we remember the current cell and add it again later\n",
      "  x,y,z = Cell\n",
      "  posorilist() = PosOriObj SimCell\n",
      "  DelObj SimCell\n",
      "  # Start an energy minimization but stop immediately.\n",
      "  # This cleans the structure in case it was supplied without hydrogens.\n",
      "  ForceField (chargefof)\n",
      "  ShowMessage 'Checking if receptor is ready for docking...'\n",
      "  Experiment Minimization\n",
      "  Experiment On\n",
      "  Experiment Off\n",
      "  # Create a receptor ensemble with flexibility corresponding to the specified temperature\n",
      "  ShowMessage 'Creating receptor ensemble with (receptors) members, each with alternative high-scoring side-chain conformations...' \n",
      "  Style Ribbon,Stick\n",
      "  ForceField YASARA2\n",
      "  Temp (temp)\n",
      "  # Don't change side-chain rotamers of residues with dative bonds to metals\n",
      "  FixRes all with arrow to metal\n",
      "  OptimizeAll Method=SCWALL,Structures=(receptors-1)\n",
      "  FreeAll\n",
      "  DelObj SimCell\n",
      "  NumberObj all,1\n",
      "  # Add original cell again\n",
      "  Cell (x),(y),(z)\n",
      "  PosOriObj SimCell,(posorilist)\n",
      "  # Optimize the hydrogen bonding network of the ensemble members\n",
      "  RemoveObj !SimCell \n",
      "  for i=1 to receptors\n",
      "    AddObj (i)\n",
      "    ShowMessage 'Optimizing hydrogen bonding network of receptor ensemble member (i)/(receptors)...'\n",
      "    OptHydAll\n",
      "    NameObj (i),Receptor(i)\n",
      "    RemoveObj (i)\n",
      "  AddObj !SimCell\n",
      "  SaveSce (ensfilename)\n",
      "  HideMessage\n",
      "# Load receptor ensemble scene\n",
      "LoadSce (ensfilename)\n",
      "# The segment name C*** is used to tag ligand Conformations, and cannot be used for the receptor\n",
      "hit = ListAtom Segment C???\n",
      "if hit\n",
      "  MarkAtom (hit)\n",
      "  segname = SegAtom (hit)\n",
      "  RaiseError 'Segment names starting with \"C\" are unfortunately reserved to identify ligand conformations, please click \"Edit > Rename > Segment\" to rename the receptor segment \"(segname)\"' \n",
      "# The molecule name 'l' is reserved for the ligand\n",
      "hit = ListAtom Mol l\n",
      "if hit\n",
      "  RaiseError 'The receptor must not contain a molecule named \"l\", please click Edit > Rename > Molecule and try again'\n",
      "# Docking is done without periodic boundaries\n",
      "Boundary Wall\n",
      "Longrange None  \n",
      "ForceField (chargefof)\n",
      "# Count receptors, in case the user provided the ensemble\n",
      "receptors = CountObj !SimCell\n",
      "runs=receptors*receptorruns\n",
      "# Loop over the receptors in the ensemble\n",
      "RemoveObj !SimCell\n",
      "for i=1 to receptors\n",
      "  AddObj (i)\n",
      "  # Show progress (ShowMessage is used by the docking experiment)\n",
      "  LabelAll 'Receptor ensemble member (i)/(receptors)',Height=0.3,Color=Yellow,Y=3.0\n",
      "  # Load the ligand for this receptor ensemble member\n",
      "  ligand=0\n",
      "  for type in 'yob','pdb'\n",
      "    filename='(MacroTarget)_ligand.(type)'\n",
      "    exists = FileSize (filename)\n",
      "    if exists \n",
      "      ligand = Load(type) (filename)\n",
      "      break\n",
      "  if not ligand\n",
      "    RaiseError 'The ligand was not found at (filename)'\n",
      "  # Just in case a user-defined ensemble was loaded\n",
      "  NameObj (i),Receptor(i)\n",
      "  # Number the ligands sequentially\n",
      "  NameObj (ligand),Ligand(i)\n",
      "  # Keep selected side-chains flexible\n",
      "  if flexres!=''\n",
      "    FixObj (i)\n",
      "    FreeAtom Res (flexres) Sidechain and Obj (i)\n",
      "    FixRes Cys Atom SG with bond to Atom SG or Ala Pro and Obj (i)\n",
      "  # Do not show secondary structure for protein ligands, this slows things down\n",
      "  HideSecStrObj Ligand(i)\n",
      "  ShowObj Ligand(i)\n",
      "  StickObj Ligand(i)\n",
      "  # Fix ligand's internal degrees of freedom if requested \n",
      "  if rigid\n",
      "    FixObj Ligand(i)\n",
      "  # Align ligand with the cell (to end up with the same local atom coordinates independent\n",
      "  # of cell orientation, which are used to calculate the checksum for the *.adr file) \n",
      "  TransferObj Ligand(i),SimCell,Local=Keep\n",
      "  # Perform the docking\n",
      "  # DEFINE HERE THE CELL!!\n",
      "  Cell Auto,extension=17, Shape=Cuboid, Res Tyr 29\n",
      "  Experiment Docking\n",
      "    Method (method)\n",
      "    ReceptorObj (i)\n",
      "    LigandObj Ligand(i)\n",
      "    Runs (receptorruns)\n",
      "    ClusterRMSD (rmsdmin)\n",
      "    # Result file number must be separated with '_', in case MacroTarget also ends with number\n",
      "    ResultFile (MacroTarget)_(i)_001\n",
      "    # Uncomment below to set the number of energy evaluations (AutoDock ga_num_evals):\n",
      "    # DockPar ga_num_evals 25000000\n",
      "    # Uncomment below to add 1000 to the random number seed when more than 999 runs are needed\n",
      "    # DockPar seed 1000\n",
      "    # Uncomment the two lines below to provide your own atom parameters (VdW radii etc.)\n",
      "    # GridPar parameter_file /Path/To/Custom/AD4_parameters.dat\n",
      "    # DockPar parameter_file /Path/To/Custom/AD4_parameters.dat\n",
      "    # Uncomment below to keep temporary files in the current working directory:\n",
      "    # TmpFileID 1adb_tmp\n",
      "  Experiment On\n",
      "  Wait ExpEnd\n",
      "  UnlabelAll\n",
      "  RemoveObj (i) Ligand(i) RecepFlex\n",
      "  NameObj RecepFlex,RecFlex(i)\n",
      "if scesaved\n",
      "  # Save a scene with all receptor and all ligand conformations\n",
      "  AddObj all\n",
      "  SaveSce (MacroTarget)\n",
      "  RemoveObj !SimCell\n",
      "\n",
      "# Collect the initial results\n",
      "Console off\n",
      "for i=1 to receptors\n",
      "  AddObj (i) Ligand(i)\n",
      "  for j=001 to receptorruns\n",
      "    # To pass the docking results back to this macro, the docking experiment\n",
      "    # stores the binding energy as the B-factor...\n",
      "    run=(i-1)*receptorruns+j\n",
      "    resultlist(run),bindnrglist(run) = DockingResult i,j,'Obj (i)','Obj Ligand(i) Segment C(j)'\n",
      "    if !(run%10)\n",
      "      ShowMessage 'Analyzing docking results, (100*run/runs)% completed...'\n",
      "      Wait 1\n",
      "  RemoveObj (i) Ligand(i)\n",
      "# Sort the results by binding energy\n",
      "ShowMessage 'Sorting docking results...'\n",
      "Wait 1\n",
      "SortResults 'bindnrglist','resultlist',''\n",
      "# Save a log file with an initial analysis\n",
      "RecordLog (MacroTarget)\n",
      "print 'Ensemble docking result analysis'\n",
      "print '================================'\n",
      "print\n",
      "print 'The ligand was docked (receptorruns) times [\"Run\"] with (method) against each of the (receptors) receptors [\"Rec\"] in the ensemble, yielding the following (runs) results, sorted by binding energy:'\n",
      "print '[More positive energies indicate stronger binding, and negative energies mean no binding]'\n",
      "print\n",
      "print 'Rec | Run |Bind.energy[kcal/mol]|Dissoc. constant [pM]| Contacting receptor residues'\n",
      "print '----+-----+---------------------+---------------------+-----------------------------'\n",
      "for i=1 to runs\n",
      "  print (resultlist(i))\n",
      "print\n",
      "\n",
      "# Get cell dimensions, needed to move the atoms in the cluster YOb files\n",
      "cd1,cd2,cd3=Cell\n",
      "# Delete previous resultlist and bindnrglist\n",
      "bindnrglist()=0\n",
      "resultlist()=0\n",
      "# Now merge the clusters from the receptor ensemble members\n",
      "clusters=0\n",
      "for i=1 to receptors\n",
      "  for j=001 to receptorruns\n",
      "    filename='(MacroTarget)_(i)_(j).yob'\n",
      "    exists = FileSize (filename)\n",
      "    if exists\n",
      "      ShowMessage 'Merging clusters from receptor ensemble member (i) run (j), (clusters) found so far...'\n",
      "      Wait 1\n",
      "      cluobj = LoadYOb (filename)\n",
      "      UnlabelAll\n",
      "      # Place the object in the dock cell to aid visualization \n",
      "      MoveAtom Obj (cluobj),(-cd/2)\n",
      "      TransferObj (cluobj),SimCell,Local=Keep  \n",
      "      result,bindnrg = DockingResult i,j,'Obj (cluobj) Segment !C???','Obj (cluobj) Segment C???'\n",
      "      added=1\n",
      "      # Keep only the ligand to save time and memory, otherwise we can quickly exceed one million atoms on screen\n",
      "      DelRes Segment !C??? Obj (cluobj)\n",
      "      if clusters and merged\n",
      "        # Get ligand RMSDs from all other clusters\n",
      "        rmsdlist() = RMSDAtom Element !H Segment C??? Obj (cluobj),Element !H Obj (join cluobjlist)  \n",
      "        for k=1 to clusters\n",
      "          if rmsdlist(k)<rmsdmin\n",
      "            # Another cluster is close, remember the one with the better binding energy\n",
      "            added=0\n",
      "            if bindnrg>bindnrglist(k)\n",
      "              # Current one has better binding energy, delete old one\n",
      "              resultlist(k)=result\n",
      "              bindnrglist(k)=bindnrg\n",
      "              filenamelist(k)=filename\n",
      "              SwapObj (cluobjlist(k)),(cluobj)\n",
      "            DelObj (cluobj)\n",
      "            break\n",
      "      if added\n",
      "        # Found a new cluster\n",
      "        clusters=clusters+1\n",
      "        resultlist(clusters)=result\n",
      "        bindnrglist(clusters)=bindnrg\n",
      "        cluobjlist(clusters)=cluobj\n",
      "        filenamelist(clusters)=filename\n",
      "ShowMessage 'Sorting cluster results...'\n",
      "Wait 1\n",
      "DelObj (join cluobjlist)\n",
      "SortResults 'bindnrglist','resultlist','filenamelist'\n",
      "# Save the unique clusters in order, mainly needed for the dock_play.mcr\n",
      "for i=1 to clusters\n",
      "  obj = LoadYOb (filenamelist(i))\n",
      "  # Transfer again into the cell\n",
      "  MoveAtom Obj (obj),(-cd/2)\n",
      "  TransferObj (obj),SimCell,Local=Keep  \n",
      "  SaveYOb (obj),(MacroTarget)_(000+i).yob\n",
      "  ShowMessage 'Saving sorted cluster (i)/(clusters)...'\n",
      "  Wait 1\n",
      "  DelObj (obj)\n",
      "# Delete original clusters on disk\n",
      "for i=1 to receptors\n",
      "  for j=001 to receptorruns\n",
      "    DelFile (MacroTarget)_(i)_(j).yob\n",
      "if scesaved\n",
      "  # Bring original objects back, unless that would exhaust memory\n",
      "  AddAll\n",
      "\n",
      "# Print cluster log\n",
      "print 'After clustering the (runs) runs, the following (clusters) distinct complex conformations were found:'\n",
      "print '[They all differ by at least (rmsdmin) A heavy ligand atom RMSD, and have been saved as YOb files with terminal number \"Num\"]' \n",
      "print\n",
      "print 'Num | Rec | Run |Bind.energy[kcal/mol]|Dissoc. constant [pM]| Contacting receptor residues'\n",
      "print '----+-----+-----+---------------------+---------------------+-----------------------------'\n",
      "for i=1 to clusters\n",
      "  print '(000+i) | (resultlist(i))'\n",
      "Style Ribbon\n",
      "StopLog\n",
      "Console On\n",
      "HideMessage\n",
      "# Exit YASARA if this macro was provided as command line argument in console mode\n",
      "if runWithMacro and ConsoleMode\n",
      "  Exit\n",
      "\n",
      "# EXTRACT DOCKING RESULT\n",
      "# ======================\n",
      "# Returns a result string and binding energy, extracted from atomic BFactor and Property.\n",
      "# 'rec' is the receptor number in the ensemble, 'run' is the docking run for this receptor, \n",
      "# 'recsel' selects the receptor, 'ligsel' selects the ligand.\n",
      "def DockingResult rec,run,recsel,ligsel\n",
      "  # Get the binding energy from the B-factor...\n",
      "  bindnrg = BFactorAtom (ligsel)\n",
      "  # ...and the dissociation constant from the atomic property.\n",
      "  dissconst = PropAtom (ligsel)\n",
      "  if bindnrg<=0\n",
      "    dissconst='               None'\n",
      "  else\n",
      "    dissconst= 00000000000000.0000+dissconst\n",
      "  # Get receptor residues that contact the ligand\n",
      "  reslist() = ListRes (recsel) with distance<4.0 from (ligsel),Format='RESNAME MOLNAME RESNUM'\n",
      "  if !count reslist\n",
      "    reslist='No residues in contact'\n",
      "  else\n",
      "    reslist=join reslist\n",
      "  result='(000+rec) | (000+run) |         (000000.0000+bindnrg) | (dissconst) | (reslist)'\n",
      "  return result,bindnrg\n",
      "\n",
      "# SORT LISTS BY VALUES, HIGHEST FIRST\n",
      "# ===================================\n",
      "# Lists are passed by reference, see Yanaconda doc section 'Calls to user-defined functions..'\n",
      "def SortResults vallist,key1list,key2list\n",
      "  caller (vallist),(key1list),(key2list)\n",
      "\n",
      "  elements=count (vallist)\n",
      "  # Create a list of indices to sort\n",
      "  for i=1 to elements\n",
      "    indexlist(i)=i\n",
      "  # Sort the list of indices  \n",
      "  passes=0\n",
      "  do\n",
      "    sorted=1\n",
      "    for i=1 to elements-1\n",
      "      if (vallist)(i)<(vallist)(i+1)\n",
      "        swap=(vallist)(i)\n",
      "        (vallist)(i)=(vallist)(i+1)\n",
      "        (vallist)(i+1)=swap\n",
      "        swap=indexlist(i)\n",
      "        indexlist(i)=indexlist(i+1)\n",
      "        indexlist(i+1)=swap\n",
      "        sorted=0\n",
      "    if elements>250 and not passes%10\n",
      "      ShowMessage 'Sorting results, pass (passes)/(elements)..'\n",
      "      Wait 1\n",
      "    passes=passes+1\n",
      "  while not sorted\n",
      "  # Rebuild the sorted keylists\n",
      "  for i=1 to 2\n",
      "    if key(i)list!=''\n",
      "      for j=1 to elements\n",
      "        keylist(j)=(key(i)list)(indexlist(j))\n",
      "      (key(i)list)()=keylist\n",
      "  HideMessage\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "macromodel = \"dock_model.mcr\"\n",
    "macro = \"dock_runensemble_dummy.mcr\"\n",
    "\n",
    "with open(macromodel) as f:\n",
    "    macrotext = f.read()\n",
    "macrotext = \"flexres = ''\".join(macrotext.split(\"# $$FLEXRES$$\"))\n",
    "macrotext = \"Cell Auto,extension=17, Shape=Cuboid, Res Tyr 29\".join(macrotext.split(\"# $$CELL$$\"))\n",
    "print(a)\n",
    "with open(macro, \"w\") as f:\n",
    "    f.write(macrotext)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "y.Exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
