{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO - --------------------------\n",
      "INFO - SEQUENCE VALIDITY CHECK:\n",
      "INFO - --------------------------\n",
      "INFO - NOTE : 0 sequences out of 1 are valid\n",
      "INFO - --------------------\n",
      "INFO - INVALID SEQUENCE(S)\n",
      "INFO - --------------------\n",
      "INFO - NOTE : The following sequence(s) contains invalid characters other than the 20 standard amino acids \n",
      "\n",
      "INFO - 1.Your input sequence\n",
      "WARNING - NOTE : ALL SEQUENCES IN YOUR FILE CONTAIN ONE OR MORE INVALID CHARACTERS !!\n",
      "WARNING - NOTE : PLEASE CHECK YOUR INPUT SEQUENCE(S) AND TRY AGAIN !\n",
      "WARNING - NOTE : PROGRAM WILL EXIT NOW !\n",
      "DEBUG - Got the results for Your input sequence\n",
      "DEBUG - 1 out of 0 complete\n"
     ]
    }
   ],
   "source": [
    "OUTPUT_FILE = \"output_yeast_structure.csv\"\n",
    "INPUT_FILE = '/home/imhof_team/Public/mauricio/workflow/yeast_type3/yeast_type3_safe2.fasta'\n",
    "\n",
    "import logging\n",
    "logging.basicConfig(level=logging.DEBUG, format='%(levelname)s - %(message)s')\n",
    "\n",
    "#import experiments\n",
    "import seqdhbm.SeqDHBM as SeqDHBM\n",
    "import seqdhbm.fasta as fasta\n",
    "import seqdhbm.pdbfiles as pdbfiles\n",
    "import sys, os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "def workflow(jobnum= \"J0\", fastafile=None, pdbs=[], pdbid=\"\", rawseq=\"\", mode=\"structure\"):\n",
    "    \"\"\" Runs the workflow \n",
    "    \n",
    "    Params:\n",
    "    - fasta: One fasta file\n",
    "    - pdbfiles: list of pdb filenames, they should be in the root dir\n",
    "    - pdbid: string with PDB codes (comma separated)\n",
    "    - knowstruc: boolean: we use services to obtain the structure if False\"\"\"\n",
    "    # TODO: use jobnum to organize submissions\n",
    "    # Deal with the fasta file with multiple sequences\n",
    "    seq_dict = {}\n",
    "    \n",
    "    seq_list = []\n",
    "    if fastafile:\n",
    "        \"\"\"seq_dict = fasta.fasta_to_seq(fastafile)\n",
    "        fasta.organize_sequences(seq_dict)\"\"\"\n",
    "        seq_list = fasta.fasta_to_seq2(fastafile)\n",
    "        fasta.organize_sequences2(seq_dict)\n",
    "        #logging.debug(\"fasta:\")\n",
    "        #logging.debug(seq_dict.keys())\n",
    "\n",
    "    # Deal with the pdb ids (by downloading them)\n",
    "    pdbid_dict = {} # stores the pdbfile path to each pdb id\n",
    "    if pdbid:\n",
    "        raise Exception(\"not prepared to work with they new variables\")\n",
    "        l = pdbfiles.text_to_list(pdbid)\n",
    "        pdbid_dict = pdbfiles.get_pdb_files(l)\n",
    "        #logging.debug(\"pdbid\")\n",
    "        #logging.debug(pdbid_dict)\n",
    "    \n",
    "    if rawseq:\n",
    "        seq_dict[\"Your input sequence\"] = (\"Your input sequence header\", rawseq)\n",
    "        folder = \"%s/MI/\"%jobnum\n",
    "        file = folder+\"MI1.fasta\"\n",
    "        name = \"Your input sequence\"\n",
    "        \"\"\"assert False, \"create the folder first\" \"\"\"\n",
    "        os.makedirs(folder, exist_ok=True)\n",
    "        with open(file, \"w\") as f:\n",
    "            f.write(\">%s\\n\"%name)\n",
    "            lines=[rawseq[x:x+60] for x in range(0, len(rawseq), 60)]\n",
    "            f.write('\\n'.join(lines))\n",
    "        seq_list += [{\"seq\":rawseq, \n",
    "                     \"name\":name, \n",
    "                     \"folder\": folder,\n",
    "                     \"file\": file,\n",
    "                     \"submited_as\": \"Manual input\"}]\n",
    "    \n",
    "    # Arrange the pdb files (user-submitted or downloaded)\n",
    "    ## Put the structure file in folders - for docking and MD\n",
    "    ## get the primary sequence for motif search\n",
    "    for file in pdbs:\n",
    "        raise Exception(\"not prepared to work with they new variables\")\n",
    "        try:\n",
    "            folder = \".\".join(file.split(\".\")[:-1]) # remove extension\n",
    "        except:\n",
    "            folder = file\n",
    "        try:\n",
    "            os.makedirs(folder, exist_ok=True)\n",
    "            shutil.move(file, folder+\"/\"+file)\n",
    "            pdbid_dict[folder] = folder+\"/\"+file\n",
    "        except:\n",
    "            logging.error(\"Could not move file\"+ file + \"\\nSkiping it!\") # TODO: List of warnings and errors\n",
    "            # TODO: organize the files into folders\n",
    "    \n",
    "    # TODO: convert the pdbs into fasta\n",
    "    for file in pdbid_dict.values():\n",
    "        seq = pdbfiles.structure_to_fasta(file)\n",
    "        if seq:\n",
    "            seq_dict[file] = (\">header\", seq)\n",
    "    # Run the motif check\n",
    "    #logging.debug(\"*\"*30)\n",
    "    #logging.debug(seq_dict)\n",
    "    \n",
    "    #spacerinfo = {}\n",
    "    \n",
    "    hbm_result = {}\n",
    "    cnt_progress = 0\n",
    "    seq_dict = {}\n",
    "    for item in seq_list:\n",
    "        cnt_progress+=1\n",
    "        try:\n",
    "            analysis = SeqDHBM.SpotCoordinationSite({\">\"+item[\"name\"]: item[\"seq\"]}, mode)\n",
    "            analysedSeq = analysis[\">\" + item[\"name\"]]\n",
    "\n",
    "            # Output? Save in our user output?\n",
    "            item[\"result\"] = analysedSeq[\"result\"]\n",
    "            item[\"warnings\"] = analysedSeq[\"warnings\"]\n",
    "\n",
    "        except Exception as e:\n",
    "            logging.error(\"****\")\n",
    "            logging.error(\"Sequence name:%s \\n %s\"%(item[\"name\"], item[\"file\"]))\n",
    "            logging.error(e)\n",
    "            logging.error(type(e))\n",
    "            logging.error(e.args)\n",
    "            logging.error(\"****\")\n",
    "        finally:\n",
    "            logging.debug(\"Got the results for %s\"%item[\"name\"])\n",
    "            logging.debug(\"%d out of %d complete\"%(cnt_progress, len(seq_dict)))\n",
    "\n",
    "    return seq_list\n",
    "\n",
    "\n",
    "if (__name__ == \"__main__\"):\n",
    "\n",
    "    fastafile = ''\n",
    "    mode = \"\"\n",
    "    if sys.argv[0].endswith(\"ipykernel_launcher.py\"):\n",
    "        fastafile = \"\"# INPUT_FILE\n",
    "        rawseq = \"ABCDZEFGH,\"\n",
    "        mode = \"structure\"\n",
    "        scriptname = \"workflow\" # TODO: Will this be the final name?\n",
    "    else:\n",
    "        scriptname = os.path.basename(__file__)\n",
    "    pdbids = \"\"# \"1S0L,1L2H,1I1B,1HIB,1I8H\"\n",
    "    pdbs=[]\n",
    "    \n",
    "    for params in sys.argv:\n",
    "        if params.startswith(\"-pdbid=\"):\n",
    "            pdbids = params[7:]\n",
    "        elif params.startswith(\"-fasta=\"):\n",
    "            fastafile = params[7:]\n",
    "        elif params.startswith(\"-pdbfiles=\"):\n",
    "            # param pdbid must be formatted as list\n",
    "            pdbs = eval(params[10:])\n",
    "        elif params.startswith(\"-mode=\"):\n",
    "            # param pdbid must be formatted as list\n",
    "            mode = params[6:].lower()\n",
    "        elif params.startswith(\"-help\"):\n",
    "            print(\"Usage Help\")\n",
    "            print(\"call by using 'python %s -mode=<mode> -fasta=<fastafile> -pdbid=<pdb_id> -pdbfiles=<pdb_files>'\"%(scriptname))\n",
    "            print(\"<mode> is mandatory and accepts:\")\n",
    "            print(\"-mode=structure: if structure is known (default)\")\n",
    "            print(\"-mode=wesa: to use our services to PREDICT surface accessibility\")\n",
    "            print(\"<fastafile> is the full path of a fasta file with one or more sequences\")\n",
    "            print(\"<pdb_id> must contain the PDB ids, separated by commas.\")\n",
    "            print(\"<pdb_files> PDB files with .ent or .pdb format\")\n",
    "    \n",
    "    assert mode in [\"structure\", \"wesa\"], (\"Mode not detected!\\n\"\n",
    "    \"Use 'python %s' for help on using the application\\n\")\n",
    "            \n",
    "    workflow(jobnum=0, fastafile=fastafile, pdbs=pdbs, rawseq=rawseq,pdbid=pdbids, mode=mode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
