{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta:\n",
      "{'271(271)': ('>IL-1A|271', 'MAKVPDMFEDLKNCYSENEEDSSSIDHLSLNQKSFYHVSYGPLHEGCMDQSVSLSISETSKTSKLTFKESMVVVATNGKVLKKRRLSLSQSITDDDLEAIANDSEEEIIKPRSAPFSFLSNVKYNFMRIIKYEFILNDALNQSIIRANDQYLTAAALHNLDEAVKFDMGAYKSSKDDAKITVILRISKTQLYVTAQDEDQPVLLKEMPEIPKTITGSETNLLFFWETHGTKNYFTSVAHPNLFIATKQDYWVCLAGGPPSITDFQILENQA'), '268(300)': ('>IL-1B|268', 'MAEVPELASEMMAYYSGNEDDLFFEADGPKQMKCSFQDDLCPLDGGIQLRISDHHYSKGFRQAASVVVAMDKLRKMLVPCPQTFQENDLSTFFPFIFEEEPIFFDTWDNEAYVHDAPVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSSAQNQWLQDMTTHLILRSFKEFLQSSLRALRQM')}\n",
      "Structure exists: '1S0L/pdb1s0l.ent' \n",
      "Structure exists: '1L2H/pdb1l2h.ent' \n",
      "Structure exists: '1I1B/pdb1i1b.ent' \n",
      "Structure exists: '1HIB/pdb1hib.ent' \n",
      "Structure exists: '1I8H/pdb1i8h.ent' \n",
      "pdbid\n",
      "{'1S0L': '1S0L/pdb1s0l.ent', '1L2H': '1L2H/pdb1l2h.ent', '1I1B': '1I1B/pdb1i1b.ent', '1HIB': '1HIB/pdb1hib.ent', '1I8H': '1I8H/pdb1i8h.ent'}\n",
      "Warning: File 1I8H/pdb1i8h.ent contains nonstandard amino acids. Skipping.\n",
      "******************************\n",
      "{'271(271)': ('>IL-1A|271', 'MAKVPDMFEDLKNCYSENEEDSSSIDHLSLNQKSFYHVSYGPLHEGCMDQSVSLSISETSKTSKLTFKESMVVVATNGKVLKKRRLSLSQSITDDDLEAIANDSEEEIIKPRSAPFSFLSNVKYNFMRIIKYEFILNDALNQSIIRANDQYLTAAALHNLDEAVKFDMGAYKSSKDDAKITVILRISKTQLYVTAQDEDQPVLLKEMPEIPKTITGSETNLLFFWETHGTKNYFTSVAHPNLFIATKQDYWVCLAGGPPSITDFQILENQA'), '268(300)': ('>IL-1B|268', 'MAEVPELASEMMAYYSGNEDDLFFEADGPKQMKCSFQDDLCPLDGGIQLRISDHHYSKGFRQAASVVVAMDKLRKMLVPCPQTFQENDLSTFFPFIFEEEPIFFDTWDNEAYVHDAPVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSSAQNQWLQDMTTHLILRSFKEFLQSSLRALRQM'), '1S0L/pdb1s0l.ent': ('>header', 'APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVWSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS'), '1L2H/pdb1l2h.ent': ('>header', 'APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVWSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNFYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS'), '1I1B/pdb1i1b.ent': ('>header', 'APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS'), '1HIB/pdb1hib.ent': ('>header', 'APVRSLNCGLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS')}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . 71(271)\n",
      "--------------------------------------------------------------------------------\n",
      "MAKVPDMFEDLKNCYSENEEDSSSIDHLSLNQKSFYHVSYGPLHEGCMDQSVSLSISETSKTSKLTFKESMVVVATNGKVLKKRRLSLSQSITDDDLEAIANDSEEEIIKPRSAPFSFLSNVKYNFMRIIKYEFILNDALNQSIIRANDQYLTAAALHNLDEAVKFDMGAYKSSKDDAKITVILRISKTQLYVTAQDEDQPVLLKEMPEIPKTITGSETNLLFFWETHGTKNYFTSVAHPNLFIATKQDYWVCLAGGPPSITDFQILENQA\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 271\n",
      "Total number of potential coordination sites found:  19\n",
      "Number of potential CYS based sites: 3\n",
      "Number of potential HIS based sites: 6\n",
      "Number of potential TYR based sites: 10 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  19 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 13 out of the 19 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 13 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 11 out of 13 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 9 out of 11 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+---------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |   Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+---------------+------------+\n",
      "|  1   |         C14          | DLKNCYSEN  | -1(CYS motif) |    YES     |\n",
      "|  2   |         Y36          | QKSFYHVSY  |       2       |     NO     |\n",
      "|  3   |         H37          | KSFYHVSYG  |       2       |     NO     |\n",
      "|  4   |         Y40          | YHVSYGPLH  |       2       |    YES     |\n",
      "|  5   |         C47          | LHEGCMDQS  | -1(CYS motif) |    YES     |\n",
      "|  6   |         Y124         | SNVKYNFMR  |       2       |    YES     |\n",
      "|  7   |         Y132         | RIIKYEFIL  |       1       |    YES     |\n",
      "|  8   |         Y171         | DMGAYKSSK  |       1       |    YES     |\n",
      "|  9   |         Y192         | KTQLYVTAQ  |       1       |    YES     |\n",
      "|  10  |         H228         | FWETHGTKN  |       1       |    YES     |\n",
      "|  11  |         Y233         | GTKNYFTSV  |       1       |    YES     |\n",
      "+------+----------------------+------------+---------------+------------+\n",
      "271(271)\n",
      "{'C14': 'YES', 'Y36': 'NO', 'H37': 'NO', 'Y40': 'YES', 'C47': 'YES', 'Y124': 'YES', 'Y132': 'YES', 'Y171': 'YES', 'Y192': 'YES', 'H228': 'YES', 'Y233': 'YES'}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . 68(300)\n",
      "--------------------------------------------------------------------------------\n",
      "MAEVPELASEMMAYYSGNEDDLFFEADGPKQMKCSFQDDLCPLDGGIQLRISDHHYSKGFRQAASVVVAMDKLRKMLVPCPQTFQENDLSTFFPFIFEEEPIFFDTWDNEAYVHDAPVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSSAQNQWLQDMTTHLILRSFKEFLQSSLRALRQM\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 300\n",
      "Total number of potential coordination sites found:  18\n",
      "Number of potential CYS based sites: 5\n",
      "Number of potential HIS based sites: 5\n",
      "Number of potential TYR based sites: 8 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  18 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "NOTE : 2 out of the 18 9mers are CP Motifs\n",
      "NOTE : These 2 9mers will skip net charge check\n",
      "-----------------------------------\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 12 out of the 18 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 12 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 11 out of 12 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 8 out of 11 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |  Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "|  1   |         C34          | KQMKCSFQD  |      1       |    YES     |\n",
      "|  2   |         H54          | RISDHHYSK  |      3       |     NO     |\n",
      "|  3   |         H55          | ISDHHYSKG  |      2       |     NO     |\n",
      "|  4   |         Y56          | SDHHYSKGF  |      2       |     NO     |\n",
      "|  5   |         C123         | RSLNCTLRD  |      1       |    YES     |\n",
      "|  6   |         H145         | LKALHLQGQ  |      2       |    YES     |\n",
      "|  7   |         C186         | LYLSCVLKD  | 0(CYS motif) |    YES     |\n",
      "|  8   |         Y205         | DPKNYPKKK  |      3       |    YES     |\n",
      "|  9   |         H280         | DMTTHLILR  |      1       |    YES     |\n",
      "|  10  |         C41          | QDDLCPLDG  | -3(CP motif) |    YES     |\n",
      "|  11  |         C80          | MLVPCPQTF  | 0(CP motif)  |    YES     |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "268(300)\n",
      "{'C34': 'YES', 'H54': 'NO', 'H55': 'NO', 'Y56': 'NO', 'C123': 'YES', 'H145': 'YES', 'C186': 'YES', 'Y205': 'YES', 'H280': 'YES', 'C41': 'YES', 'C80': 'YES'}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . S0L/pdb1s0l.ent\n",
      "--------------------------------------------------------------------------------\n",
      "APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVWSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 153\n",
      "Total number of potential coordination sites found:  7\n",
      "Number of potential CYS based sites: 2\n",
      "Number of potential HIS based sites: 1\n",
      "Number of potential TYR based sites: 4 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  7 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 6 out of the 7 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 6 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 4 out of 6 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 4 out of 4 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |  Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "|  1   |          C8          | RSLNCTLRD  |      1       |    YES     |\n",
      "|  2   |         H30          | LKALHLQGQ  |      2       |    YES     |\n",
      "|  3   |         C71          | LYLSCVLKD  | 0(CYS motif) |    YES     |\n",
      "|  4   |         Y90          | DPKNYPKKK  |      3       |    YES     |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "1S0L/pdb1s0l.ent\n",
      "{'C8': 'YES', 'H30': 'YES', 'C71': 'YES', 'Y90': 'YES'}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . L2H/pdb1l2h.ent\n",
      "--------------------------------------------------------------------------------\n",
      "APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVWSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNFYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 153\n",
      "Total number of potential coordination sites found:  7\n",
      "Number of potential CYS based sites: 2\n",
      "Number of potential HIS based sites: 1\n",
      "Number of potential TYR based sites: 4 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  7 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 6 out of the 7 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 6 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 4 out of 6 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 4 out of 4 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |  Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "|  1   |          C8          | RSLNCTLRD  |      1       |    YES     |\n",
      "|  2   |         H30          | LKALHLQGQ  |      2       |    YES     |\n",
      "|  3   |         C71          | LYLSCVLKD  | 0(CYS motif) |    YES     |\n",
      "|  4   |         Y90          | DPKNYPKKK  |      3       |    YES     |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "1L2H/pdb1l2h.ent\n",
      "{'C8': 'YES', 'H30': 'YES', 'C71': 'YES', 'Y90': 'YES'}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . I1B/pdb1i1b.ent\n",
      "--------------------------------------------------------------------------------\n",
      "APVRSLNCTLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 153\n",
      "Total number of potential coordination sites found:  7\n",
      "Number of potential CYS based sites: 2\n",
      "Number of potential HIS based sites: 1\n",
      "Number of potential TYR based sites: 4 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  7 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 6 out of the 7 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 6 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 4 out of 6 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 4 out of 4 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |  Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "|  1   |          C8          | RSLNCTLRD  |      1       |    YES     |\n",
      "|  2   |         H30          | LKALHLQGQ  |      2       |    YES     |\n",
      "|  3   |         C71          | LYLSCVLKD  | 0(CYS motif) |    YES     |\n",
      "|  4   |         Y90          | DPKNYPKKK  |      3       |    YES     |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "1I1B/pdb1i1b.ent\n",
      "{'C8': 'YES', 'H30': 'YES', 'C71': 'YES', 'Y90': 'YES'}\n",
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . HIB/pdb1hib.ent\n",
      "--------------------------------------------------------------------------------\n",
      "APVRSLNCGLRDSQQKSLVMSGPYELKALHLQGQDMEQQVVFSMSFVQGEESNDKIPVALGLKEKNLYLSCVLKDDKPTLQLESVDPKNYPKKKMEKRFVFNKIEINNKLEFESAQFPNWYISTSQAENMPVFLGGTKGGQDITDFTMQFVSS\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 153\n",
      "Total number of potential coordination sites found:  7\n",
      "Number of potential CYS based sites: 2\n",
      "Number of potential HIS based sites: 1\n",
      "Number of potential TYR based sites: 4 \n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the  7 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 6 out of the 7 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 6 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 4 out of 6 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 4 out of 4 PASS the spacer check!\n",
      "NOTE : All checks done, preparing tabular summary\n",
      "**************************************************************************************************************\n",
      "TABULAR SUMMARY\n",
      "NOTE : Please use the available structure information to only consider those motifs that are \"surface exposed\"\n",
      "**************************************************************************************************************\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "| S.no | Coordinating residue | 9mer motif |  Net charge  | Spacer > 2 |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "|  1   |          C8          | RSLNCGLRD  |      1       |    YES     |\n",
      "|  2   |         H30          | LKALHLQGQ  |      2       |    YES     |\n",
      "|  3   |         C71          | LYLSCVLKD  | 0(CYS motif) |    YES     |\n",
      "|  4   |         Y90          | DPKNYPKKK  |      3       |    YES     |\n",
      "+------+----------------------+------------+--------------+------------+\n",
      "1HIB/pdb1hib.ent\n",
      "{'C8': 'YES', 'H30': 'YES', 'C71': 'YES', 'Y90': 'YES'}\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "logging.basicConfig(level=logging.DEBUG, format='%(levelname)s - %(message)s')\n",
    "\n",
    "#import experiments\n",
    "import SeqDHBM\n",
    "import fasta\n",
    "import pdbfiles\n",
    "import sys, os\n",
    "import shutil\n",
    "\n",
    "def workflow(jobnum= 0, fastafile=None, pdbs=[], pdbid=\"\", mode=True):\n",
    "    \"\"\" Runs the workflow \n",
    "    \n",
    "    Params:\n",
    "    - fasta: One fasta file\n",
    "    - pdbfiles: list of pdb filenames, they should be in the root dir\n",
    "    - pdbid: string with PDB codes (comma separated)\n",
    "    - knowstruc: boolean: we use services to obtain the structure if False\"\"\"\n",
    "    # TODO: use jobnum to organize submissions\n",
    "    # Deal with the fasta file with multiple sequences\n",
    "    if fastafile:\n",
    "        seq_dict = fasta.fasta_to_seq(fastafile)\n",
    "        fasta.organize_sequences(seq_dict)\n",
    "        print(\"fasta:\")\n",
    "        print(seq_dict)\n",
    "\n",
    "    # Deal with the pdb ids (by downloading them)\n",
    "    pdbid_dict = {} # stores the pdbfile path to each pdb id\n",
    "    if pdbid:\n",
    "        l = pdbfiles.text_to_list(pdbid)\n",
    "        pdbid_dict = pdbfiles.get_pdb_files(l)\n",
    "        print(\"pdbid\")\n",
    "        print(pdbid_dict)\n",
    "    \n",
    "    # Arrange the pdb files (user-submitted or downloaded)\n",
    "    ## Put the structure file in folders - for docking and MD\n",
    "    ## get the primary sequence for motif search\n",
    "    for file in pdbs:\n",
    "        try:\n",
    "            folder = \".\".join(file.split(\".\")[:-1]) # remove extension\n",
    "        except:\n",
    "            folder = file\n",
    "        try:\n",
    "            os.makedirs(folder, exist_ok=True)\n",
    "            shutil.move(file, folder+\"/\"+file)\n",
    "            pdbid_dict[folder] = folder+\"/\"+file\n",
    "        except:\n",
    "            logging.error(\"Could not move file\"+ file + \"\\nSkiping it!\") # TODO: List of warnings and errors\n",
    "            # TODO: organize the files into folders\n",
    "    \n",
    "    # TODO: convert the pdbs into fasta\n",
    "    for file in pdbid_dict.values():\n",
    "        seq = pdbfiles.structure_to_fasta(file)\n",
    "        if seq:\n",
    "            seq_dict[file] = (\">header\", seq)\n",
    "    # Run the motif check\n",
    "    print(\"*\"*30)\n",
    "    print(seq_dict)\n",
    "    \n",
    "    spacerinfo = {}\n",
    "    \n",
    "    for folder, (header, seq) in seq_dict.items():\n",
    "        \"\"\"fo = \"Saved_run/IL-1A(271)/IL-1A(271).fasta\"\n",
    "    \n",
    "        SeqDHBM.ReadFasta(fo)\"\"\"\n",
    "        #print(folder, \"$\", seq)\n",
    "        coordsites = SeqDHBM.SpotCoordinationSite({folder: seq}, mode)\n",
    "        for folder, coordsite in coordsites.items():\n",
    "            # TODO: Handle what is to be done with the coordination sites\n",
    "            # Output? Save in our user output?\n",
    "            print(folder)\n",
    "            print(coordsite)\n",
    "\n",
    "if (__name__ == \"__main__\"):\n",
    "\n",
    "    fastafile = '/home/imhof_team/Public/mauricio/workflow/test.fasta'\n",
    "    pdbids = \"1S0L,1L2H,1I1B,1HIB,1I8H\"\n",
    "    pdbs=[]\n",
    "    mode = \"structure\"\n",
    "    \n",
    "    for params in sys.argv:\n",
    "        if params.startswith(\"-fasta=\"):\n",
    "            fastafile = params[7:]\n",
    "    \n",
    "    for params in sys.argv:\n",
    "        if params.startswith(\"-pdbid=\"):\n",
    "            pdbids = params[7:]\n",
    "\n",
    "    for params in sys.argv:\n",
    "        if params.startswith(\"-pdbfiles=\"):\n",
    "            # param pdbid must be formatted as list\n",
    "            pdbs = eval(params[10:])\n",
    "            \n",
    "    for params in sys.argv:\n",
    "        if params.startswith(\"-mode=\"):\n",
    "            # param pdbid must be formatted as list\n",
    "            mode = params[6:].lower()\n",
    "    \n",
    "    assert mode in [\"structure\", \"wesa\"], (\"Please choose one of the modes: \\n\"\n",
    "    \"- structure: if structure is known (default)\\n\"\n",
    "    \"- wesa: to use our services to PREDICT surface accessibility\\n\")\n",
    "            \n",
    "    workflow(jobnum=0, fastafile=fastafile, pdbs=pdbs, pdbid=pdbids, mode=mode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG - This is a log message.1\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "logging.debug('This is a log message.'+str(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
