{% load static from staticfiles %}
<!DOCTYPE html>
<html>
<head>
  {% load staticfiles %}
  <link rel="icon" href="{% static 'favicon.ico' %}">
<style>
body{
  background-color: paleturquoise ;
}
h1 {
  color: black;
  text-align: center;
}
p.center {
  text-align: justify;
  color: black;
}
<style>
table, th, td {
    border: 1px solid black;
}
</style> <!-- TODO Put the style for the table in a CSS -->
<!--
Examples for website style:
http://pipe.scs.fsu.edu/wesa/
http://www.cbs.dtu.dk/services/NetSurfP/
https://zhanglab.ccmb.med.umich.edu/I-TASSER/
-->

</head>
<body>
<h1>[Seq]uence based [D]etection of [H]eme [B]inding [M]otifs</h1>
<p class="center">
 This webpage provides temporary access to our implementation of SeqD-HBM, a computational approach for the prediction of transient heme binding motifs from protein sequences.
</p>
<p class="center">
  The program processes the input sequence through a systematic stepwise validation,
  extracting at each step relevant features from the sequence to produce a tabulated list of the possible heme-coordination sites available for the given sequence.
</p>
<img src="{% static 'SeqDHBM/algotithm_flow.png' %}"/>

<p class="center">
If you use SeqD-HBM in any part of your work, please cite:<br>
1. Wißbrock, A., Paul George, A.A., Brewitz, H. H., Kühl, T., Imhof, D. (2019) The Molecular Basis of Transient Heme-Protein Interactions: Analysis, Concept and Implementation. <a href="http://www.bioscirep.org/content/39/1/BSR20181940"><i>Biosci. Rep. 39(1), BSR20181940</i></a><br><br>
If you use <a href="http://pipe.scs.fsu.edu/wesa/">WESA</a> to perform solvent accessibility checks on your sequence please cite the relevant literature as well.<br>
1. Chen, H-L and Zhou, H-X. Prediction of solvent accessibility and sites of deleterious mutations from protein sequence. Nucl. Acids Res. 33, 3193-3199 (2005).<br>
2. Shan, Y., Wang, G., and Zhou, H.-X. Fold recognition and accurate query-template alignment by a combination of PSI-BLAST and threading. Proteins 42, 23-37 (2001).<br>
</p>
<p class="center">
Please Note : Sequences exceeding 2000 residues will not be processed via WESA but the usual SeqD-HBM analysis will be done. Moreover, depending on sequence length, each sequence requires anything between 5 to 30 minutes for processing by WESA.
</p>
<p class="center">
  Further enhancements to the web interface and the underlying logic are a work in progress.
  For queries regarding SeqD-HBM and the usage of this site please contact Ajay A. Paul George (ajgeorge@uni-bonn.de) 
</p>
<p class="center">
 If our research interests you, please visit our <a href="https://www.pharma.uni-bonn.de/www/pharmazeutische-biochemie-und-bioanalytik">homepage</a> for further details.
</p>
<!--p>
  <b>References:</b>
<ul>
 <li>Tony's paper on motif importance (<a href="brokenlink">question</a>)</li>
</ul>
</p-->
  <!--p>
    Cite WESA</p-->
<hr>
<h3>SeqD-HBM User Input:</h3>
<form enctype="multipart/form-data" action="/SeqDHBM/" method="post" autocomplete="off">
  {% csrf_token %}
  {% if errors %}
  <p>
    <!-- https://docs.djangoproject.com/en/2.1/ref/forms/api/#how-errors-are-displayed -->
    And the errors are:
    {{ errors.as_p }}
  </p>
  {% endif %}
  {% for field in form %}
      <div class="fieldWrapper">
          <p> {{ field.label_tag }} </p>
          <p> {{ field }} </p>
          <font color="fa0000"> {{ field.errors }} </font>
      </div>
  {% endfor %}
  <input type="submit" value="Submit">
</form>

<p id="userresult"></p>
{% if results %}
    {% for result in results %}
        <div>
          <H3>  {{ result.name }} </H3>
          <p><font size="-1" face= 'Courier New'> {{ result.seq }} </font> </p>
          {% for coord, data in result.result.items %}
              {{ coord }}
              {{ data.ninemer }}
              {{ data.netcharge }}
              {{ data.comment }}
          {% endfor %}
          {% for warn in result.warnings %}
            <p> {{ warn }}</p>
          {% endfor %}
        </div>
    {% endfor %}
{% endif %}
<font color="00fa00">{{ debug }}</font>
</body>
</html>
