<html>
<head>
<style>
table, th, td {
    border: 1px solid black;
}
</style> <!-- TODO Put the style for the table in a CSS -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script>
$(document).ready(function(){
    // replace this for the commented code after implementing submission of emails
    $("#run").click(function(){
      $.get("runworkflow",
      {
          //pdbids:document.getElementById('pdbids').value,
          //fastafile:document.getElementById('fastafile').value,
          aaseq:document.getElementById('aaseq').value,
          //pdbfiles:document.getElementById('pdbfiles').value,
          mode:$("input[name=mode]:checked").val()

      },
      function (data, status) {
          document.getElementById('userresult').innerHTML = data
      });
      // replace until here


      /*if ((document.getElementById('email').value == "") && ($("input[name=mode]:checked").val()=="wesa")) {
        document.getElementById('emailwarn').innerHTML = "This field is mandatory for structure predictions";
      } else {
        document.getElementById('emailwarn').innerHTML = ""
        $.get("runworkflow",
        {
            //pdbids:document.getElementById('pdbids').value,
            //fastafile:document.getElementById('fastafile').value,
            aaseq:document.getElementById('aaseq').value,
            //pdbfiles:document.getElementById('pdbfiles').value,
            mode:$("input[name=mode]:checked").val()

        },
        function (data, status) {
            document.getElementById('userresult').innerHTML = data
        });
      }*/
  });
});
</script>
</head>
<body>
<h1>[Seq]uence based [D]etection of [H]eme [B]inding [M]otifs</h1>
<p>
  The term "heme-binding motif" (HBM) is used to describe a protein sequence stretch that may interact with the heme molecule.
  The evaluation of the binding behavior of a plethora of heme-peptide complexes resulted in the implementation of a prediction tool (SeqD-HBM) for heme-binding
      motifs which eventually led and will perspectively lead to the identification and verification of so far unknown-heme-regulated proteins.
</p>
<p>
  A handful of computational tools are available to predict ligand binding to proteins. While these tools use complex algorithms and base their predictions on
   extracting several novel features, the nature of binding usually described is the strong irreversible binding of heme (mostly within a binding pocket) and
   not the transient binding of and regulation by heme.
</p>
<p>
  The program processes the input sequence through a systematic stepwise validation,
  extracting at each step relevant features from the sequence to produce a tabulated list of the possible heme-coordination sites available for the given sequence.
</p>
<p>
  Besides reporting the potential heme-coordination sites, the program additionally outputs the 9mer motifs associated with the coordination site,
  the charge of the motif and the possibility of a disulfide bridge close to the coordinating atom of the 9mer.
  This comprehensive output further guides the user to make informed judgments on the nature of heme interaction with the protein of interest.
</p>
<p>
  Further improvements to the interface are a work in progress.
</p>
<!--p>
  <b>References:</b>
<ul>
 <li>Tony's paper on motif importance (<a href="brokenlink">question</a>)</li>
</ul>
</p>
  <!--p>
    Cite WESA</p-->
<hr>
<h3>SeqD-HBM Online service:</h3>

<!--
Examples for website design:
http://pipe.scs.fsu.edu/wesa/
http://www.cbs.dtu.dk/services/NetSurfP/
https://zhanglab.ccmb.med.umich.edu/I-TASSER/

{% if variablename %}
    <ul>
    {% for question in variablename %}
        <li><a href="/polls/{{ question }}/">{{ question }}</a></li>
    {% endfor %}
    </ul>
{% else %}
    <p>No variables are available.</p>
{% endif %}
<hl> <!--Disabling controls that will first be used in future versions -->

<!--Hiding controls that will first be used in future versions
<p>Enter your email address to receive the complete analysis of the submitted sequences.
  As the structure prediction can take a long time (from 5 to 90 minutes), this is mandatory if you want the structure to be predicted:</p>
<p id="emailwarn" style="color:red;"></p>
<p><input disabled id="email" size="50" type="text" /></p>
<p>Fasta file:</p>
<p><input id="fastafile" size="50" type="file" /></p>
-->
<p>Paste your sequence data:</p>
<p><textarea id="aaseq" cols="80" rows="4"></textarea></p>
 <!--Hiding controls that will first be used in future versions
<p>PDB files:</p>
<p><input id="pdbfiles" type="file" /></p>
<p>PDB ids:</p>
<p><input id="pdbids" size="80" type="text"/></p>
-->
<p>Filter prediction according to solvent accessibility (WESA):</p>
<!--
# for wesa prediction: read
https://stackoverflow.com/questions/1239035/asynchronous-method-call-in-python
-->
<div>
<div><input checked="checked" disabled id="mode" name="mode" type="radio" value="structure" /> <label for="structure">Skip</label></div>
<div><input disabled id="mode" name="mode" type="radio" value="wesa" /> <label for="wesa">Filter</label></div>
</div>
<p><input id="run" name="send" type="submit" value="Run" /></p>

<p id="userresult"></p>

</body>
</html>
