{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#/usr/bin/env python\n",
    "# Import Modules and Clear the windows command prompt screen\n",
    "import sys\n",
    "import time\n",
    "import datetime\n",
    "import os\n",
    "import os.path\n",
    "import pathlib\n",
    "import shutil\n",
    "from prettytable import PrettyTable\n",
    "import logging\n",
    "if __name__ == \"__main__\":\n",
    "    logging.basicConfig(level=logging.DEBUG, format='%(message)s')\n",
    "else:\n",
    "    logging.basicConfig(level=logging.ERROR, format='%(levelname)s - %(message)s')\n",
    "\n",
    "start = time.time() # Program start time \n",
    "#os.system('cls')\n",
    "#os.system('clear')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Auxiliary functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CalcExecTime(start):\n",
    "    \"\"\" \"\"\"\n",
    "    logging.info(\"\\n\")\n",
    "    logging.info(\"*\" * 80)\n",
    "    end = round((time.time() - start),2)\n",
    "    logging.info(\"Execution time: \"+ str(end)+ \" seconds\")\n",
    "    logging.info(\"*\" * 80)\n",
    "    \n",
    "\n",
    "def ReadFasta(filename):\n",
    "    \"\"\"1.Read the contents from the fasta file\n",
    "    2.Count the number of headers\n",
    "    3.logging.info the number of sequences\n",
    "    4.Store the header and sequence in a dictionary as key value pairs \n",
    "    5.Return the dictionary\"\"\"\n",
    "    fasta_dict={}                           # Initialize an empty dictionary\n",
    "    content_list=[]                         # Read the entire content of the file into this list\n",
    "    header_list=[]                          # Initialize an empty list to collect the headers in the file\n",
    "    header_index_list=[]                    # Initialize an empty list to collect the indices headers in the file\n",
    "    sequence_list=[]                        # Initialize an empty list to collect the contents under each header\n",
    "    with open(filename, 'r') as infile:\n",
    "        for line in infile:\n",
    "            line = line.rstrip()\n",
    "            if line:                            # To avoid empty lines from being added to the list\n",
    "                content_list.append(line)       # populate the list, each line from the fasta file is each element of the list\n",
    "    counter = 0             # Initialize a counter to spot header indices since list.index() function does not work like in version 2.7\n",
    "    for element in content_list:\n",
    "        counter += 1\n",
    "        if element[0] == \">\":                           # Spot header lines \n",
    "            header_index = counter - 1                  # Stores the index of the headers \n",
    "            header_list.append(content_list[header_index]) # populate the header list\n",
    "            header_index_list.append(header_index)\n",
    "    \n",
    "    for i in range(len(header_index_list)): # Iterate through the no. of headers i.e. the no. of sequences\n",
    "        if i !=(len(header_index_list)-1):  # Check that we are not at the last header yet\n",
    "            startval = header_index_list[i]+1 # Starting of list range \n",
    "            endval = header_index_list[i+1] # Ending of list range\n",
    "            sequence_element =(\"\".join(content_list[startval:endval])) # Join the lines between startval and endval\n",
    "            sequence_list.append(sequence_element) # each element is an individual sequence\n",
    "        \n",
    "        if i ==(len(header_index_list)-1): # For the last iteration\n",
    "            startval = header_index_list[i]+1\n",
    "            endval = len(content_list)+1\n",
    "            sequence_element =(\"\".join(content_list[startval:endval]))\n",
    "            sequence_list.append(sequence_element)\n",
    "        \n",
    "            \"\"\" Header and sequence list created \"\"\"\n",
    "            \"\"\" Now store header and sequence as key value pairs in the fasta_dict {} \"\"\"\n",
    "    key = 0\n",
    "    for header in header_list:\n",
    "        fasta_dict[header] = sequence_list[key] \n",
    "        key += 1\n",
    "    logging.info(\"--------------------------\")\n",
    "    logging.info(\"INPUT FILE CONTENT SUMMARY:\")\n",
    "    logging.info(\"--------------------------\")\n",
    "    logging.info(\"NOTE : Your input file contains \"+str(len(header_list))+\" sequence(s)\")\n",
    "    return(fasta_dict) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Important checks in case the program is called from command line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "********************************************************************************\n",
      "Seq-HBM v0.9\n",
      "SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\n",
      "Thursday , October-11-2018, 16:19:10\n",
      "********************************************************************************\n",
      "\n",
      "INPUT FILE MISSING ! EXITING NOW !\n",
      "\n",
      "\n",
      "********************************************************************************\n",
      "Execution time: 1.43 seconds\n",
      "********************************************************************************\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/imhof_team/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2918: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "if (__name__ == \"__main__\"):\n",
    "    if (sys.argv[0][-21:] == \"ipykernel_launcher.py\"):\n",
    "        # running from notebook for testing\n",
    "        scriptname = \"Seq-HBM\"\n",
    "        logging.info('*'*80)\n",
    "        logging.info(scriptname+\" v0.11\")\n",
    "        logging.info(\"SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\")\n",
    "        logging.info(datetime.datetime.today().strftime(\"%A , %B-%d-%Y, %H:%M:%S\"))\n",
    "        logging.info('*'*80)\n",
    "        input_file_path = \"/home/imhof_team/Public/mauricio/SeqD-HBM\"\n",
    "        input_file_name = \"test.fasta\"\n",
    "        operation_mode = \"structure\"\n",
    "        if(os.path.isfile(os.path.join(input_file_path,input_file_name))): #Check if the file exists\n",
    "            fo=os.path.join(input_file_path,input_file_name) \n",
    "            logging.info(\"NOTE : Your input file is: \"+ input_file_name)\n",
    "        else:\n",
    "            logging.info(\"\\nINPUT FILE MISSING ! EXITING NOW !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()\n",
    "    else:\n",
    "        ##################################################\n",
    "        scriptname = os.path.basename(__file__)\n",
    "        logging.info('*'*80)\n",
    "        logging.info(scriptname+\" v0.11\")\n",
    "        logging.info(\"SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\")\n",
    "        logging.info(datetime.datetime.today().strftime(\"%A , %B-%d-%Y, %H:%M:%S\"))\n",
    "        logging.info('*'*80)\n",
    "        #################################################\n",
    "        if (len(sys.argv)<3 or (sys.argv[2] not in [\"wesa\",\"structure\"])):\n",
    "            logging.error(\"TOO FEW or INCORRECT command line arguments !\")\n",
    "            logging.error(\"The program needs 2 command line arguments as follows\")\n",
    "            logging.error(\"python \"+scriptname+\" Argument1 Argument2\")\n",
    "            logging.error(\"Where..\")\n",
    "            logging.error(\"Argument1 : Input file name with the full file extension\")\n",
    "            logging.error(\"Argument2 : Operation mode, either \\\"wesa\\\" or \\\"structure\\\"\")\n",
    "            logging.error(\"Please try again !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()\n",
    "        #File Path\n",
    "        input_file_path = os.path.dirname(os.path.realpath(__file__))\n",
    "        logging.info(input_file_path)\n",
    "        input_file_name=sys.argv[1]\n",
    "        operation_mode=sys.argv[2]\n",
    "\n",
    "        if(os.path.isfile(os.path.join(input_file_path,input_file_name))): #Check if the file exists\n",
    "            fo=os.path.join(input_file_path,input_file_name)\n",
    "            logging.info(\"NOTE : Your input file is: \"+ input_file_name)\n",
    "        else:\n",
    "            logging.error(\"\\nINPUT FILE MISSING ! EXITING NOW !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Program functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################################################################################################################\n",
    "################################################\n",
    "#           Function: SequenceValidityCheck()\n",
    "################################################\n",
    "#   1.Take a list of characters and check if each element in the list is part of the 20 standard amino acids\n",
    "#   2.Return a flag value based on the number of errors in the input\n",
    "#   3.If all characters are valid then the flag should be 0 and 0 is returned\n",
    "##################################################          \n",
    "def SequenceValidityCheck(current_sequence_list):\n",
    "    standard_aa_list = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y'] # List of the 20 standard amino acids\n",
    "    flag = 0\n",
    "    for i in range(len(current_sequence_list)):\n",
    "        if current_sequence_list[i] not in standard_aa_list:\n",
    "            flag +=1\n",
    "    return(flag)\n",
    "\n",
    "\n",
    "################################################################################################################################\n",
    "################################################\n",
    "#           Function: BuildNinemerSlice()\n",
    "################################################\n",
    "#   1.Take a sequence as string and a index value to indicate the coordinating amino acid as int\n",
    "#   2.Check if the index is within 3 amino acids from either of the terminals\n",
    "#   3.If found in the terminals add apporiate number of amino acids on the opposite sides to form the 9mer \n",
    "#   4.Return the 9 mer as string\n",
    "##################################################          \n",
    "def BuildNinemerSlice(seq,ind):\n",
    "    seq_length = len(seq)\n",
    "    \"\"\"When the coordinating residue is found in the first 5 residues of the sequence\"\"\"\n",
    "    if(ind <= 4): # When the coordinating residue is found in the first 5 residues of the sequence\n",
    "        ninemer = seq[0:9] # String slice of the first 9 AAs of the sequence\n",
    "        return(ninemer)\n",
    "    \"\"\"When the coordinating residue is found in the last 5 residues of the sequence\"\"\"\n",
    "    if(ind >= seq_length-5): # When the coordinating residue is found in the last 5 residues of the sequence\n",
    "        ninemer = seq[seq_length-9:seq_length] # String slice of the last 9 AAs of the sequence\n",
    "        return(ninemer)\n",
    "    if(ind >= 5 and ind < seq_length-5): # When the coordinating residue is in between and outside the above two conditions\n",
    "        ninemer = seq[ind-4:ind+5] # String slice of the last 4 AAs from either side of the coord AA\n",
    "        return(ninemer)\n",
    "\n",
    "\n",
    "##############################################################################################################################\n",
    "################################################\n",
    "#           Function: CheckBasicAdjacent()\n",
    "################################################\n",
    "#   1.Take a string of characters which is the 9mer as input\n",
    "#   2.Check if there are basic amino acids in the 9mer other than the coodrinating AA\n",
    "#   3.Return success or failure \n",
    "##################################################\n",
    "def CheckBasicAdjacent(ninemer_sequence):\n",
    "    basic_aa_list = ['R', 'K', 'H'] # List of basic and positively charged amino acids\n",
    "    flag = \"n\"\n",
    "    for i in range(len(ninemer_sequence)):\n",
    "        if(i != 4): # Skip iterating on the coordinating AA in the sequence\n",
    "            if(ninemer_sequence[i] in basic_aa_list):\n",
    "                flag = \"y\"\n",
    "                return(flag)\n",
    "    return(flag)\n",
    "\n",
    "def CheckCPMotif(coordinating_aa):\n",
    "    \"\"\"Checks if there is a Proline adjacent to the Cysteine.\n",
    "    \n",
    "    Returns Boolean\"\"\"\n",
    "    return (\"CP\" in ninemer_sequence) or (\"PC\" in ninemer_sequence)\n",
    "    \n",
    "##############################################################################################################################\n",
    "################################################\n",
    "#           Function: CheckNetCharge()\n",
    "################################################\n",
    "#   1.Take a string of characters which is the 9mer as input\n",
    "#   2.Compute the net charge on the motif\n",
    "#   3.Return success or failure\n",
    "##################################################\n",
    "def CheckNetCharge(ninemer_sequence):\n",
    "    basic_aa_list = ['R', 'K', 'H'] # List of basic and positively charged amino acids\n",
    "    neg_aa_list = ['D','E'] # List of negatively charged amino acids\n",
    "    net_charge = 0\n",
    "    for i in range(len(ninemer_sequence)):\n",
    "        if(ninemer_sequence[i] in basic_aa_list):\n",
    "            net_charge += 1\n",
    "        if(ninemer_sequence[i] in neg_aa_list):\n",
    "            net_charge -= 1\n",
    "    return(net_charge)\n",
    "\n",
    "################################################\n",
    "#           Function: BuildWESANinemerSlice()\n",
    "################################################\n",
    "#   1.Take a sequence as string and a index value to indicate the coordinating amino acid as int\n",
    "#   2.Check if the index is within 3 amino acids from either of the terminals\n",
    "#   3.If found in the terminals add apporiate number of amino acids on the opposite sides to form the 9mer \n",
    "#   4.Return the 9 mer as string\n",
    "##################################################          \n",
    "def ShipSeqToWESA(seq,coord_list):\n",
    "    if sys.argv[0].endswith(\"ipykernel_launcher.py\"):\n",
    "        basepath=\"/home/imhof_team/Public/mauricio/workflow\"\n",
    "    else:\n",
    "        basepath=os.path.dirname(os.path.realpath(__file__))\n",
    "    wesa_tmpdir=os.path.join(basepath,'WESA_tmp')\n",
    "    \n",
    "    if os.path.exists(wesa_tmpdir):\n",
    "        shutil.rmtree(wesa_tmpdir) # Delete previously created WESA_tmp directory and all its contents \n",
    "    if not os.path.exists(wesa_tmpdir):\n",
    "        os.makedirs(wesa_tmpdir)\n",
    "        shutil.copy2('WESA-submit.py',wesa_tmpdir) # Copy this script into the WESA_tmp directory\n",
    "        shutil.copy2('MultipartPostHandler.py',wesa_tmpdir) # Copy this library into the WESA_tmp directory\n",
    "        if os.path.isfile('wget.exe'):\n",
    "            shutil.copy2('wget.exe',wesa_tmpdir) # Copy wget command into the tmp directory for windows\n",
    "        os.chdir(wesa_tmpdir)\n",
    "        with open('seq_for_wesa.txt', 'w') as f: # Open a text file named seq_for_wesa.txt and write the sequence into it\n",
    "            f.write(seq)\n",
    "        jobname=\"SeqDHBM2WESA\"\n",
    "        email=\"mauriciopl@gmail.com\"\n",
    "        inp_file=\"seq_for_wesa.txt\"\n",
    "        html=jobname+'.html'\n",
    "        wesa_tup='python2 WESA-submit.py',jobname, email, inp_file, '>', html # This is stored as a tuple\n",
    "        wesa_str=' '.join(wesa_tup) # Convert the tpule to string\n",
    "        #logging.info(wesa_str)\n",
    "        os.system(wesa_str) # Send the job to the WESA server\n",
    "        out_list=\"\"\n",
    "        wesa_wget_tup='wget -O',jobname+'.out', '-F -i', html\n",
    "        wesa_wget_str=' '.join(wesa_wget_tup) # Convert the tpule to string\n",
    "        #logging.info(wesa_wget_tup[1])\n",
    "        os.system(wesa_wget_str) # Download the output as a .out file\n",
    "        file_size=os.stat(wesa_wget_tup[1]) # contains the file size of the output file\n",
    "        initial_file_size=file_size.st_size\n",
    "        logging.info(\"*\"*80)\n",
    "        logging.info(\"NOTE : Your sequence has been posted to the WESA server for solvent accessibility prediction\")\n",
    "        logging.info(\"NOTE : This might take a few minutes....grab a coffee maybe?\")\n",
    "        logging.info(\"NOTE : Making attempts at 30 second intervals to see if the WESA prediction is complete\")\n",
    "        logging.info(\"*\"*80)\n",
    "        final_file_size=0\n",
    "        attempt_count=0\n",
    "        while (initial_file_size>=final_file_size):\n",
    "            attempt_count+=1\n",
    "            logging.info(\"NOTE : Attempt \" + str(attempt_count)+\" to fetch WESA output...\")\n",
    "            logging.info(\"*\"*80)\n",
    "            time.sleep(30)\n",
    "            os.system(wesa_wget_str) # Download and overwrite the .out file\n",
    "            time.sleep(2)\n",
    "            f_size=os.stat(wesa_wget_tup[1])\n",
    "            final_file_size=f_size.st_size\n",
    "        logging.info(\"*\"*80)\n",
    "        logging.info(\"NOTE : WESA solvent accessibility prediction complete !\")\n",
    "        logging.info(\"*\"*80)\n",
    "\n",
    "        with open('SeqDHBM2WESA.out', 'r') as f: # Read the contents of the output file from WESA\n",
    "            line_list=[]\n",
    "            out_list=[]\n",
    "            for line in f:\n",
    "                line = line.lstrip() # Strip spaces from either side \n",
    "                line = line.rstrip()\n",
    "                line_list.append(line)\n",
    "            for i in range(len(coord_list)):\n",
    "                for j in range(len(line_list)):\n",
    "                    if (line_list[j].startswith(str(coord_list[i])+' ')): \n",
    "                        out_list.append(line_list[j][-7:-6])\n",
    "        os.chdir(\"..\")\n",
    "    return(out_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################################################################################################################################\n",
    "################################################\n",
    "#           Function: SpotCoordinationSite()\n",
    "#############################################################################################################################################################\n",
    "#   1.Take the dictionary fasta_dict from the ReadFasta() function\n",
    "#   2.Check sequence validity by checking if they are alphabets from the set of the one letter codes of the 20 standard amino acids via SequenceValidityCheck()\n",
    "#   3.Take the first record and logging.infothe number of coordination sites found \n",
    "#   4.logging.infothe number of C, H and Y and their position in the sequence\n",
    "#   5.Build the 9mer motif for each coordination site via BuildNinemerSlice()\n",
    "#   5.For each 9mer motif check(a) Adjacent basic amino acids via CheckBasicAdjacent()\n",
    "#                              (b) Net charge in the 9mer motif via CheckNetCharge()\n",
    "##############################################################################################################################################################\n",
    "def SpotCoordinationSite(fasta_dict, mode):\n",
    "    filtered_dict = {} # Empty dictionary that will contain only valid sequences\n",
    "    invalid_seq_dict  = {} # Empty dictionary to collect all invalid seqence headers\n",
    "    sno = 0\n",
    "    coordination_site_count = 0\n",
    "    seq_with_coord = 0\n",
    "    MAX_SEQ_LENGTH_FOR_WESA = 2000\n",
    "    output = {} # the results of the check. Key is the fasta header and the values are a \n",
    "                # dictionary with {coordinating_atom: spacercheck}\n",
    "    logging.info(\"--------------------------\")\n",
    "    logging.info(\"SEQUENCE VALIDITY CHECK:\")\n",
    "    logging.info(\"--------------------------\")\n",
    "    for header, sequence in fasta_dict.items(): # Iterate through the dictionary\n",
    "        sno += 1\n",
    "        current_sequence =(sequence)\n",
    "        current_sequence_list = list(current_sequence) # Split the sequence into a list\n",
    "\n",
    "        #logging.info(\"***** Checking sequence\",sno,\"under header:\", header[1:],\"*****\")\n",
    "        if(SequenceValidityCheck(current_sequence_list) == 0): # Pass the list to the SequenceValidityCheck function\n",
    "            #logging.info(\"NOTE : Sequence content check: SEQUENCE VALID !\") # if all sequences have standard amino acids in the sequences then the value will be 0\n",
    "            filtered_dict[header] = sequence # Add only the valid sequences to the new dictionary\n",
    "        else:\n",
    "            invalid_seq_dict[header] = sequence\n",
    "    if (len(fasta_dict) == len(filtered_dict)):\n",
    "        logging.info(\"NOTE : All the \"+str(len(fasta_dict))+\" sequences were verified to be valid !\")\n",
    "    else:\n",
    "        logging.info(\"NOTE : \"+str(len(filtered_dict))+\" sequences out of \"+str(len(fasta_dict))+\" are valid\")\n",
    "        logging.info(\"--------------------\")\n",
    "        logging.info(\"INVALID SEQUENCE(S)\")\n",
    "        logging.info(\"--------------------\")\n",
    "        logging.info(\"NOTE : The following sequence(s) contains invalid characters other than the 20 standard amino acids \\n\")\n",
    "        inv=0\n",
    "        for head in invalid_seq_dict:\n",
    "            inv += 1\n",
    "            logging.info(\"%d.%s\"%(inv,head))\n",
    "\n",
    "    if(len(filtered_dict) == 0): # if this dictionary is empty, then all sequences have errors and the program must exit\n",
    "        logging.warning(\"NOTE : ALL SEQUENCES IN YOUR FILE CONTAIN ONE OR MORE INVALID CHARACTERS !!\")\n",
    "        logging.warning(\"NOTE : PLEASE CHECK YOUR INPUT SEQUENCE(S) AND TRY AGAIN !\")\n",
    "        logging.warning(\"NOTE : PROGRAM WILL EXIT NOW !\")\n",
    "        #CalcExecTime(start)\n",
    "        return dict() ## sys.exit() # Mauricio Changed\n",
    "\n",
    "    #logging.info(\"-\" * 80) \n",
    "    #logging.info(\"NOTE : The valid sequences will now be screened for the potential heme coordination sites(C, H, Y)\")\n",
    "    #logging.info(\"-\" * 80)\n",
    "    sno = 0 \n",
    "    for header in filtered_dict: # Iterate through the dictionary\n",
    "        initial_NinemerDict = {} # An empty dictionary that will contain the residue number and the associated 9 mer as key value pair eg. {H150: 'XXXXHXXXX'}\n",
    "        dict_index_site = {} # Dictionary that stores the index of a possible coordination site and the amino acid(C or H or Y)\n",
    "        coordination_site_count = 0 # This counter is used later to determine if the program must end\n",
    "        basic_adj_list = [] # Initialize an empty list to collect the flags when the check for basic adjacent amoni acids is done. Values will be either yes or no.\n",
    "        net_charge_list = [] # Initialize a list to store the net charges computed by the CheckNetCharge() function\n",
    "        pos_charge_list = [] # list with the index of 9mers with positive net charge\n",
    "        dict_for_netCharge_calc = {} # 9mer motifs that PASS the basic adjacency test are populated here\n",
    "        neg_charge_list =[]\n",
    "        pos_check_lst = []\n",
    "        pass_charge_dict = {} # contains 9 mers that have passed the net positive charge check\n",
    "        pass_charge_index_list = [] # This list contains the indices of the coordination sites after the charge check is done\n",
    "        spacer_dict = {} # Dict containing \"y\" or \"n\" values to check if two coordination sites are separated by a spacer of at least 2\n",
    "        count_basic_adj_ninemer = 0 # Initialize the count of basic adjacent AAs in each ninemer\n",
    "        pass_charge_out_list = []\n",
    "        wesa_substr_index_list = [] # List containing indices of the starting positions of the 9mers for comparison with the WESA output\n",
    "        out_coord_atoms = {} # The output of the function for the current sequence\n",
    "        cys_count = 0\n",
    "        his_count = 0\n",
    "        tyr_count = 0\n",
    "        sno += 1\n",
    "        current_sequence =(filtered_dict[header])\n",
    "        #logging.info(\"\\n\")\n",
    "        logging.info(\"-\" * 80)\n",
    "        logging.info(\"-\" * 80)\n",
    "        logging.info(\"WORKING ON THE VALID SEQUENCE \"+str(sno)+\" . \"+header[1:]) \n",
    "        logging.info(\"-\" * 80)\n",
    "        logging.info(current_sequence)\n",
    "        logging.info(\"-\" * 30)\n",
    "        logging.info(\"HEME COORDINATION SITE CHECK:\")\n",
    "        logging.info(\"-\" * 30)\n",
    "        \n",
    "        current_sequence_length = len(current_sequence)\n",
    "        cys_count = current_sequence.count('C')\n",
    "        his_count = current_sequence.count('H')\n",
    "        tyr_count = current_sequence.count('Y')\n",
    "        coordination_site_count = cys_count + his_count  + tyr_count\n",
    "    \n",
    "        if(coordination_site_count > 0):\n",
    "            seq_with_coord += 1\n",
    "            logging.info(\"NOTE : Heme coordination check PASS!\")\n",
    "            logging.info(\"Length of sequence: \"+ str(current_sequence_length))\n",
    "            logging.info(\"Total number of potential coordination sites found: \"+str(coordination_site_count))\n",
    "            logging.info(\"Number of potential CYS based sites: \"+str(cys_count))\n",
    "            logging.info(\"Number of potential HIS based sites: \"+str(his_count))\n",
    "            logging.info(\"Number of potential TYR based sites: \"+str(tyr_count)+\"\\n\")\n",
    "            \n",
    "            coord_site_index_list =([pos for pos, char in enumerate(current_sequence) if char in('H','C','Y')]) # List with the indices of the coordination sites\n",
    "            for i in range(len(coord_site_index_list)): # Iterate through the list containing the indices of the coordination sites\n",
    "                pos_on_sequence = coord_site_index_list[i] + 1 # Position on the sequence = index + 1 since Python index starts at 0\n",
    "                coord_aa = current_sequence[coord_site_index_list[i]] # Variable stores the value on the index which is the actual amino acid residue\n",
    "                formatted_coord_site =(coord_aa+str(pos_on_sequence)) # Variable stores the amino acid and index in the format eg. H151\n",
    "                ninemer = BuildNinemerSlice(current_sequence,coord_site_index_list[i]) # Call the BuildNinemerSlice function\n",
    "                initial_NinemerDict[formatted_coord_site] = ninemer # Populate the initial_NinemerDict {} in the format eg. {H12: XXXXHXXXX}\n",
    "\n",
    "            if(i ==(len(initial_NinemerDict)-1)): # Final record of the initial_NinemerDict{}\n",
    "                logging.info(\"NOTE : Successfully built 9mer motifs for all the \"+str(coordination_site_count)+\" potential coordination sites !\")\n",
    "                logging.info(\"-\" * 50)\n",
    "\n",
    "\n",
    "            logging.info(\"CP MOTIF CHECK:\")\n",
    "            logging.info(\"-\" * 35)\n",
    "            logging.info(\"NOTE : Screening all Cysteine Proline motifs\")\n",
    "            \n",
    "            \n",
    "            cp_motif = {} # This dictionary will skip net charge check\n",
    "            for aa, ninemer in initial_NinemerDict.items():\n",
    "                # check if the coordinating aa is Cysteine\n",
    "                if aa[0] == 'C':\n",
    "                    pos = int(aa[1:])-1 # the position on string\n",
    "                    if pos>0 and current_sequence[pos-1]=='P': # previous is Proline\n",
    "                        cp_motif[aa] = ninemer\n",
    "                    if pos<len(current_sequence)-2 and current_sequence[pos+1]=='P': # next is Proline\n",
    "                        cp_motif[aa] = ninemer\n",
    "\n",
    "            if len(cp_motif):\n",
    "                logging.info(\"NOTE : \"+str(len(cp_motif))+\" out of the \"+str(coordination_site_count)+\" 9mers are CP Motifs\")\n",
    "                logging.info(\"NOTE : These \"+str(len(cp_motif))+\" 9mers will skip net charge check\")\n",
    "                logging.info(\"-\" * 35)\n",
    "\n",
    "                \n",
    "            logging.info(\"ADJACENT BASIC AMINO ACID CHECK:\")\n",
    "            logging.info(\"-\" * 35)\n",
    "            logging.info(\"NOTE : Screening all potential motifs for adjacent basic amino acids\")\n",
    "            \n",
    "            for key, ninemer in initial_NinemerDict.items():\n",
    "                basic_adj_list.append(CheckBasicAdjacent(ninemer)) # this list contains values \"y\" or \"n\"\n",
    "                if (CheckBasicAdjacent(ninemer) == \"y\") and (key not in cp_motif.keys()):\n",
    "                    dict_for_netCharge_calc[key] = ninemer\n",
    "                    \n",
    "            count_basic_adj_ninemer = (basic_adj_list.count(\"y\"))\n",
    "\n",
    "            if(count_basic_adj_ninemer > 0):\n",
    "                logging.info(\"NOTE : \"+str(count_basic_adj_ninemer)+ \" out of the \"+str(coordination_site_count)+\" 9mers PASS the adjacent basic amino acids check\")\n",
    "                logging.info(\"NOTE : These \"+str(count_basic_adj_ninemer)+\" 9mers will be used in the next step to check for positive net charge\")\n",
    "\n",
    "            if(count_basic_adj_ninemer+len(cp_motif) == 0): # If all elements in the list are \"n\" this means that none of the 9mers have adjacent basic amino acids \n",
    "                logging.info(\"-\" * 50)\n",
    "                logging.info(\"NOTE : None of the 9mers have valid motifs\")\n",
    "                # logging.info(\"NOTE : This also means that none of these 9mer motifs will have a net positive charge\") OR CP\n",
    "                logging.info(\"NOTE : The likelyhood of any part of this sequence binding or coordinating heme is very low !!!\")\n",
    "                logging.info(\"NOTE : Checking the next valid sequence\")\n",
    "                logging.info(\"-\" * 50)\n",
    "                logging.info(\"-\" * 50)\n",
    "                continue\n",
    "            \n",
    "            for k,v in cp_motif.items(): \n",
    "                dict_for_netCharge_calc[k] = v\n",
    "                \n",
    "            #NET CHARGE CHECK\n",
    "            logging.info(\"-\" * 35)\n",
    "            logging.info(\"NET CHARGE CHECK:\")\n",
    "            logging.info(\"-\" * 35)\n",
    "            logging.info(\"NOTE : Checking net charge on the individual 9mer motifs\")\n",
    "            logging.info(\"NOTE : An individual motif PASSES this check if its net charge is positive\")\n",
    "            logging.info(\"NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\")\n",
    "            #logging.info(\"-\" * 50)\n",
    "            for z in dict_for_netCharge_calc:\n",
    "                net_charge_list.append(CheckNetCharge(dict_for_netCharge_calc[z]))\n",
    "            nt_cg_counter = 0 # counter to loop into the net_charge_list\n",
    "            for d in dict_for_netCharge_calc:\n",
    "                if d in cp_motif:\n",
    "                    pass_charge_dict[d] = dict_for_netCharge_calc[d]\n",
    "                    pass_charge_out_list.append (str(net_charge_list[nt_cg_counter])+\"(CP motif)\")\n",
    "                elif(net_charge_list[nt_cg_counter] > 0 ):\n",
    "                    pass_charge_dict[d] = dict_for_netCharge_calc[d]\n",
    "                    #logging.info(\"Motif\",d,\"PASSES NET CHARGE CHECK: REASON - Positive net charge:\",net_charge_list[nt_cg_counter])\n",
    "                    pass_charge_out_list.append(net_charge_list[nt_cg_counter])\n",
    "                else:\n",
    "                    if(net_charge_list[nt_cg_counter] <= 0 and d[0]==\"C\"):\n",
    "                        pass_charge_dict[d] = dict_for_netCharge_calc[d]\n",
    "                        #logging.info(\"Motif\",d,\"PASSES NET CHARGE CHECK: REASON - CYS based motif\")\n",
    "                        pass_charge_out_list.append (str(net_charge_list[nt_cg_counter])+\"(CYS motif)\")\n",
    "                nt_cg_counter += 1\n",
    "            logging.info(\"NOTE : \"+str(len(pass_charge_dict))+\" out of \"+str(count_basic_adj_ninemer)+\" PASS the net charge check!\")\n",
    "            #Additional coordination site check\n",
    "            logging.info(\"-\" * 50)\n",
    "            logging.info(\"ADDITIONAL COORDINATION SITE CHECK\")\n",
    "            logging.info(\"-\" * 50)\n",
    "            logging.info(\"NOTE : Checking the presence of additional coordination sites\")\n",
    "            if(len(pass_charge_dict) > 1):\n",
    "                logging.info(\"NOTE : Additional coordination site check: PASS\")\n",
    "                logging.info(\"NOTE : Proceeding to check spacer length between coordination sites\")\n",
    "            elif(len(pass_charge_dict) == 1):\n",
    "                logging.info(\"NOTE : No additional coordination sites found !\")\n",
    "                logging.info(\"NOTE : Spacer check will be skipped !\")\n",
    "            else:\n",
    "                logging.info(\"NOTE : No coordination sites passed the net charge check !\")\n",
    "                logging.info(\"NOTE : It is very likely that this sequence does not bind/coordinate heme !\")\n",
    "                continue\n",
    "            #logging.info(\"\\n\")\n",
    "            logging.info(\"-\" * 50)\n",
    "            logging.info(\"SPACER DISTANCE CHECK\")\n",
    "            logging.info(\"-\" * 50)\n",
    "            for head in pass_charge_dict: # Iterate through this dictionary\n",
    "                pass_charge_index_list.append(int(head[1:])) # append to the pass_charge_index_list only the number from the dictionary header eg. in C23 extract 23 and convert it to 'int'\n",
    "            pass_charge_index_list.sort()\n",
    "            if (len(pass_charge_index_list) > 1):\n",
    "                for i in range(len(pass_charge_index_list)):\n",
    "                    if (i == 0):\n",
    "                        if(abs((pass_charge_index_list[i] - pass_charge_index_list[i+1])) > 2): \n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"YES\"\n",
    "                        else:\n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"NO\"\n",
    "                    elif (i == len(pass_charge_index_list)-1):\n",
    "                        if((abs(pass_charge_index_list[i] - pass_charge_index_list[i-1])) > 2):\n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"YES\"\n",
    "                        else:\n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"NO\"\n",
    "                    else:\n",
    "                        if ((abs(pass_charge_index_list[i] - pass_charge_index_list[i+1])) > 2 and(abs(pass_charge_index_list[i] - pass_charge_index_list[i-1])) > 2):\n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"YES\"\n",
    "                        else:\n",
    "                            spacer_dict[pass_charge_index_list[i]] = \"NO\"\n",
    "            else:\n",
    "                logging.info(\"NOTE : Only one 9mer to check meaning only one valid coordination site\")\n",
    "                logging.info(\"NOTE : Spacer check will be skipped\")\n",
    "                spacer_dict[int(pass_charge_index_list[0])] = \"NO\" # position of the only valid aa in the chain is NO\n",
    "            spacer_check_pass = list(spacer_dict.values()).count(\"YES\")\n",
    "            logging.info(\"NOTE : \"+str(spacer_check_pass)+\" out of \"+str( len(pass_charge_index_list))+ \" PASS the spacer check!\")\n",
    "            \n",
    "            # WESA supports only sequences with length up to 2000\n",
    "            if (mode==\"wesa\") and (current_sequence_length>MAX_SEQ_LENGTH_FOR_WESA):\n",
    "                logging.warning(\"-\" * 50)\n",
    "                logging.warning(\"Warning:\")\n",
    "                logging.warning(\"Structure prediction by the WESA server supports sequences with\"+\n",
    "                                \" length up to %d amino acids\"%MAX_SEQ_LENGTH_FOR_WESA)\n",
    "                logging.warning(\"The current sequence has %d a.a. and is not supported.\")\n",
    "                \n",
    "            if (mode==\"structure\") or current_sequence_length>MAX_SEQ_LENGTH_FOR_WESA:\n",
    "                logging.info(\"NOTE : All checks done, preparing tabular summary\")\n",
    "                \"\"\"Prepare output lists for PrettyTable\"\"\"\n",
    "                logging.info(\"*\" * 110)\n",
    "                logging.info(\"TABULAR SUMMARY\")\n",
    "                logging.info(\"NOTE : Please use the available structure information to only consider those motifs that are \\\"surface exposed\\\"\")\n",
    "                logging.info(\"*\" * 110)\n",
    "                table = PrettyTable([\"S.no\", \"Coordinating residue\", \"9mer motif\", \"Net charge\", \"Spacer > 2\"])\n",
    "                sr_no = 0\n",
    "                for coord in pass_charge_dict:\n",
    "                    sr_no+= 1\n",
    "                    out_coord_atoms[coord] = spacer_dict[int(coord[1:])]\n",
    "                    table.add_row([sr_no, coord, pass_charge_dict[coord], pass_charge_out_list[sr_no-1], spacer_dict[int(coord[1:])]])\n",
    "                logging.info(table)\n",
    "                output[header] = out_coord_atoms\n",
    "                #logging.info(len(pass_charge_dict),len(pass_charge_out_list),len(spacer_list))\n",
    "                #return pass_charge_dict\n",
    "\n",
    "\n",
    "            if (mode==\"wesa\") and (current_sequence_length<=MAX_SEQ_LENGTH_FOR_WESA):\n",
    "                logging.info(\"Sending sequence to the WESA server for solvent accessibility prediction\")\n",
    "                wesa_return=ShipSeqToWESA(current_sequence,pass_charge_index_list) # Call the function to send the sequence to WESA passing the sequence as argument\n",
    "                pass_wesa_index_list=[]\n",
    "                wesa_spacer_dict={}\n",
    "                counter=0\n",
    "                for coord in pass_charge_dict:\n",
    "                    if(wesa_return[counter]==\"1\"):\n",
    "                        pass_wesa_index_list.append(int(coord[1:]))\n",
    "                    counter+=1\n",
    "                if (len(pass_wesa_index_list) > 1):\n",
    "                    for i in range(len(pass_wesa_index_list)):\n",
    "                        if(i == 0):\n",
    "                            if(abs((pass_wesa_index_list[i] - pass_wesa_index_list[i+1])) > 2): \n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"YES\"\n",
    "                            else:\n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"NO\"\n",
    "                        elif(i == len(pass_wesa_index_list)-1):\n",
    "                            if((abs(pass_wesa_index_list[i] - pass_wesa_index_list[i-1])) > 2):\n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"YES\"\n",
    "                            else:\n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"NO\"\n",
    "                        else:\n",
    "                            if((abs(pass_wesa_index_list[i] - pass_wesa_index_list[i+1])) > 2 and(abs(pass_wesa_index_list[i] - pass_wesa_index_list[i-1])) > 2):\n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"YES\"\n",
    "                            else:\n",
    "                                wesa_spacer_dict[pass_wesa_index_list[i]] = \"NO\"\n",
    "                else:\n",
    "                    logging.info(\"NOTE : Only one 9mer to check meaning only one valid coordination site\")\n",
    "                    logging.info(\"NOTE : Spacer check will be skipped\")\n",
    "                    spacer_dict[pass_wesa_index_list[0]] = \"NO\" # position of the only valid aa in the chain is NO\n",
    "                table = PrettyTable([\"S.no\", \"Coordinating residue\", \"9mer motif\", \"Net charge\", \"Spacer > 2\"])\n",
    "                sr_no=0\n",
    "                counter=0\n",
    "                for coord in pass_charge_dict:\n",
    "                    if(wesa_return[counter]==\"1\"):\n",
    "                        sr_no+= 1\n",
    "                        table.add_row([sr_no, coord, pass_charge_dict[coord], pass_charge_out_list[counter], wesa_spacer_dict[int(coord[1:])]])\n",
    "                        out_coord_atoms[coord] = wesa_spacer_dict[int(coord[1:])]\n",
    "                    counter+=1\n",
    "                output[header] = out_coord_atoms\n",
    "                if (sr_no>0):\n",
    "                    logging.info(table)\n",
    "                else:\n",
    "                    logging.info(\"*\"*80)\n",
    "                    logging.info(\"NOTE : YOUR SEQUENCE HAS NO SOLVENT ACCESSIBLE COORDINATION RESIDUES !\")\n",
    "                    logging.info(\"*\"*80)\n",
    "        else: # When the number of coordination sites in the sequence is 0\n",
    "            if(len(filtered_dict) == 1): # if the file contains only one sequence\n",
    "                logging.info(\"-\" * 80)\n",
    "                logging.info(\"NOTE : The sequence does not have potential heme coordination sites(C, H, Y) !!\")\n",
    "                logging.info(\"NOTE : It is very likely that this sequence does not bind/coordinate heme !\")\n",
    "                logging.info(\"NOTE : This is the only sequence on the file...nothing more to check !\")\n",
    "                logging.info(\"-\" * 80)\n",
    "                logging.info(\"PROGRAM WILL EXIT NOW !!\")\n",
    "                return output ## sys.exit() \n",
    "            else: # When the file contains more than one sequence and we are now at the last sequence\n",
    "                if(sno == len(filtered_dict)):\n",
    "                    logging.info(\"-\" * 80)\n",
    "                    logging.info(\"NOTE : It is very likely that this sequence does not bind/coordinate heme !\")\n",
    "                    logging.info(\"NOTE : The sequence does not have potential heme coordination sites(C, H, Y) !!\")\n",
    "                    logging.info(\"NOTE : This is the last sequence on the file...nothing more to check !!\")\n",
    "                    logging.info(\"-\" * 80)\n",
    "                    logging.info(\"PROGRAM WILL EXIT NOW !!\")\n",
    "                    # CalcExecTime(start)\n",
    "                    return output ## sys.exit() \n",
    "                logging.info(\"-\" * 80) # Any sequence before the last sequence\n",
    "                logging.info(\"NOTE : It is very likely that this sequence does not bind/coordinate heme\")\n",
    "                logging.info(\"NOTE : The sequence does not have potential heme coordination sites(C, H, Y), checking the next valid sequence !\")\n",
    "                logging.info(\"NOTE : Nothing more to check..moving to the next valid sequece !!\")\n",
    "                logging.info(\"-\" * 80)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run the code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "********************************************************************************\n",
      "Seq-HBM v0.10\n",
      "SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\n",
      "Thursday , October-11-2018, 16:19:29\n",
      "********************************************************************************\n",
      "\n",
      "INPUT FILE MISSING ! EXITING NOW !\n",
      "\n",
      "\n",
      "********************************************************************************\n",
      "Execution time: 20.54 seconds\n",
      "********************************************************************************\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/imhof_team/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2918: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "if (__name__ == \"__main__\"):\n",
    "    if (sys.argv[0][-21:] == \"ipykernel_launcher.py\"):\n",
    "        # running from notebook for testing\n",
    "        scriptname = \"Seq-HBM\"\n",
    "        logging.info('*'*80)\n",
    "        logging.info(scriptname+\" v0.11\")\n",
    "        logging.info(\"SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\")\n",
    "        logging.info(datetime.datetime.today().strftime(\"%A , %B-%d-%Y, %H:%M:%S\"))\n",
    "        logging.info('*'*80)\n",
    "        input_file_path = \"/home/imhof_team/Public/mauricio/SeqD-HBM\"\n",
    "        input_file_name = \"test.fasta\"\n",
    "        operation_mode = \"structure\"\n",
    "        if(os.path.isfile(os.path.join(input_file_path,input_file_name))): #Check if the file exists\n",
    "            fo=os.path.join(input_file_path,input_file_name)\n",
    "            logging.info(\"NOTE : Your input file is: \"+ input_file_name)\n",
    "        else:\n",
    "            logging.error(\"\\nINPUT FILE MISSING ! EXITING NOW !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()\n",
    "    else:\n",
    "        ##################################################\n",
    "        scriptname = os.path.basename(__file__)\n",
    "        logging.info('*'*80)\n",
    "        logging.info(scriptname+\" v0.11\")\n",
    "        logging.info(\"SeqD-HBM : [Seq]uence based [D]etection of [H]eme [B]inding [M]otifs\")\n",
    "        logging.info(datetime.datetime.today().strftime(\"%A , %B-%d-%Y, %H:%M:%S\"))\n",
    "        logging.info('*'*80)\n",
    "        #################################################\n",
    "        if (len(sys.argv)<3 or (sys.argv[2] not in [\"wesa\",\"structure\"])):\n",
    "            logging.error(\"TOO FEW or INCORRECT command line arguments !\")\n",
    "            logging.error(\"The program needs 2 command line arguments as follows\")\n",
    "            logging.error(\"python \"+scriptname+\" Argument1 Argument2\")\n",
    "            logging.error(\"Where..\")\n",
    "            logging.error(\"Argument1 : Input file name with the full file extension\")\n",
    "            logging.error(\"Argument2 : Operation mode, either \\\"wesa\\\" or \\\"structure\\\"\")\n",
    "            logging.error(\"Please try again !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()\n",
    "        #File Path\n",
    "        input_file_path = os.path.dirname(os.path.realpath(__file__))\n",
    "        logging.info(input_file_path)\n",
    "        input_file_name=sys.argv[1]\n",
    "        operation_mode=sys.argv[2]\n",
    "\n",
    "        if(os.path.isfile(os.path.join(input_file_path,input_file_name))): #Check if the file exists\n",
    "            fo=os.path.join(input_file_path,input_file_name)\n",
    "            logging.info(\"NOTE : Your input file is: \"+input_file_name)\n",
    "        else:\n",
    "            logging.error(\"\\nINPUT FILE MISSING ! EXITING NOW !\")\n",
    "            CalcExecTime(start)\n",
    "            sys.exit()\n",
    "    \n",
    "    SpotCoordinationSite(ReadFasta(fo) , operation_mode)\n",
    "    CalcExecTime(start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--------------------------\n",
      "SEQUENCE VALIDITY CHECK:\n",
      "--------------------------\n",
      "NOTE : All the 1 sequences were verified to be valid !\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "WORKING ON THE VALID SEQUENCE 1 . older\n",
      "--------------------------------------------------------------------------------\n",
      "MSKAVGIDLGTTYSCVAHFANDRVEIIANDQGNRTTPSYVAFTDTERLIGDAAKNQAAMNPHNTVFDAKRLIGRKFDDPEVTNDAKHYPFKVIDKGGKPVVQVEYKGETKTFTPEEISSMILTKMKETAENFLGTEVKDAVVTVPAYFNDSQRQATKDAGTIAGLNVLRIINEPTAAAIAYGLDKKSQKEHNVLIFDLGGGTFDVSLLSIDEGVFEVKATAGDTHLGGEDFDSRLVNFLAEEFKRKNKKDLTTNQRSLRRLRTAAERAKRTLSSSAQTSIEIDSLFEGIDFYTSITRARFEELCADLFRSTLEPVEKVLADSKLDKSQIDEIVLVGGSTRIPKVQKLVSDFFNGKEPNRSINPDEAVAYGAAVQAAILTGDQSSTTQDLLLLDVAPLSLGIETAGGIMTKLIPRNSTIPTKKSEVFSTYADNQPGVLIQVFEGERTRTKDNNLLGKFELSGIPPAPRGVPQIEVTFDIDANGILNVSAVEKGTGKSNKITITNDKGRLSKEDIDKMVAEAEKFKAEDEQEAQRVQAKNQLESYAFTLKNSVSENNFKEKVGEEDARKLEAAAQDAINWLDASQAASTEEYKERQKELEGVANPIMSKFYGAAGGAPGAGPVPGAGAGPTGAPDNGPTVEEVD\n",
      "------------------------------\n",
      "HEME COORDINATION SITE CHECK:\n",
      "------------------------------\n",
      "NOTE : Heme coordination check PASS!\n",
      "Length of sequence: 642\n",
      "Total number of potential coordination sites found: 19\n",
      "Number of potential CYS based sites: 2\n",
      "Number of potential HIS based sites: 5\n",
      "Number of potential TYR based sites: 12\n",
      "\n",
      "NOTE : Successfully built 9mer motifs for all the 19 potential coordination sites !\n",
      "--------------------------------------------------\n",
      "CP MOTIF CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all Cysteine Proline motifs\n",
      "ADJACENT BASIC AMINO ACID CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Screening all potential motifs for adjacent basic amino acids\n",
      "NOTE : 8 out of the 19 9mers PASS the adjacent basic amino acids check\n",
      "NOTE : These 8 9mers will be used in the next step to check for positive net charge\n",
      "-----------------------------------\n",
      "NET CHARGE CHECK:\n",
      "-----------------------------------\n",
      "NOTE : Checking net charge on the individual 9mer motifs\n",
      "NOTE : An individual motif PASSES this check if its net charge is positive\n",
      "NOTE : An individual motif also PASSES this check if its net charge is NOT positive BUT it is a CYS BASED motif\n",
      "NOTE : 5 out of 8 PASS the net charge check!\n",
      "--------------------------------------------------\n",
      "ADDITIONAL COORDINATION SITE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : Checking the presence of additional coordination sites\n",
      "NOTE : Additional coordination site check: PASS\n",
      "NOTE : Proceeding to check spacer length between coordination sites\n",
      "--------------------------------------------------\n",
      "SPACER DISTANCE CHECK\n",
      "--------------------------------------------------\n",
      "NOTE : 3 out of 5 PASS the spacer check!\n",
      "Sending sequence to the WESA server for solvent accessibility prediction\n",
      "********************************************************************************\n",
      "NOTE : Your sequence has been posted to the WESA server for solvent accessibility prediction\n",
      "NOTE : This might take a few minutes....grab a coffee maybe?\n",
      "NOTE : Making attempts at 30 second intervals to see if the WESA prediction is complete\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 1 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 2 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 3 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 4 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 5 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 6 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 7 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 8 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 9 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 10 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 11 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 12 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 13 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 14 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 15 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 16 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 17 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 18 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 19 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 20 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 21 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 22 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 23 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 24 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NOTE : Attempt 25 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 26 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 27 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 28 to fetch WESA output...\n",
      "********************************************************************************\n",
      "********************************************************************************\n",
      "NOTE : Attempt 29 to fetch WESA output...\n",
      "********************************************************************************\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-23ccf4713a60>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mSpotCoordinationSite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"folder\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"MSKAVGIDLGTTYSCVAHFANDRVEIIANDQGNRTTPSYVAFTDTERLIGDAAKNQAAMNPHNTVFDAKRLIGRKFDDPEVTNDAKHYPFKVIDKGGKPVVQVEYKGETKTFTPEEISSMILTKMKETAENFLGTEVKDAVVTVPAYFNDSQRQATKDAGTIAGLNVLRIINEPTAAAIAYGLDKKSQKEHNVLIFDLGGGTFDVSLLSIDEGVFEVKATAGDTHLGGEDFDSRLVNFLAEEFKRKNKKDLTTNQRSLRRLRTAAERAKRTLSSSAQTSIEIDSLFEGIDFYTSITRARFEELCADLFRSTLEPVEKVLADSKLDKSQIDEIVLVGGSTRIPKVQKLVSDFFNGKEPNRSINPDEAVAYGAAVQAAILTGDQSSTTQDLLLLDVAPLSLGIETAGGIMTKLIPRNSTIPTKKSEVFSTYADNQPGVLIQVFEGERTRTKDNNLLGKFELSGIPPAPRGVPQIEVTFDIDANGILNVSAVEKGTGKSNKITITNDKGRLSKEDIDKMVAEAEKFKAEDEQEAQRVQAKNQLESYAFTLKNSVSENNFKEKVGEEDARKLEAAAQDAINWLDASQAASTEEYKERQKELEGVANPIMSKFYGAAGGAPGAGPVPGAGAGPTGAPDNGPTVEEVD\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'wesa'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-5-67143e03ee32>\u001b[0m in \u001b[0;36mSpotCoordinationSite\u001b[0;34m(fasta_dict, mode)\u001b[0m\n\u001b[1;32m    268\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"wesa\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mcurrent_sequence_length\u001b[0m\u001b[0;34m<=\u001b[0m\u001b[0mMAX_SEQ_LENGTH_FOR_WESA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    269\u001b[0m                 \u001b[0mlogging\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Sending sequence to the WESA server for solvent accessibility prediction\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 270\u001b[0;31m                 \u001b[0mwesa_return\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mShipSeqToWESA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcurrent_sequence\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpass_charge_index_list\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# Call the function to send the sequence to WESA passing the sequence as argument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    271\u001b[0m                 \u001b[0mpass_wesa_index_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    272\u001b[0m                 \u001b[0mwesa_spacer_dict\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-9-ae331ea505ca>\u001b[0m in \u001b[0;36mShipSeqToWESA\u001b[0;34m(seq, coord_list)\u001b[0m\n\u001b[1;32m    136\u001b[0m             \u001b[0mlogging\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"NOTE : Attempt \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mattempt_count\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\" to fetch WESA output...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    137\u001b[0m             \u001b[0mlogging\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"*\"\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m80\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 138\u001b[0;31m             \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m30\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    139\u001b[0m             \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwesa_wget_str\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# Download and overwrite the .out file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "#SpotCoordinationSite({\"folder\": \"MSKAVGIDLGTTYSCVAHFANDRVEIIANDQGNRTTPSYVAFTDTERLIGDAAKNQAAMNPHNTVFDAKRLIGRKFDDPEVTNDAKHYPFKVIDKGGKPVVQVEYKGETKTFTPEEISSMILTKMKETAENFLGTEVKDAVVTVPAYFNDSQRQATKDAGTIAGLNVLRIINEPTAAAIAYGLDKKSQKEHNVLIFDLGGGTFDVSLLSIDEGVFEVKATAGDTHLGGEDFDSRLVNFLAEEFKRKNKKDLTTNQRSLRRLRTAAERAKRTLSSSAQTSIEIDSLFEGIDFYTSITRARFEELCADLFRSTLEPVEKVLADSKLDKSQIDEIVLVGGSTRIPKVQKLVSDFFNGKEPNRSINPDEAVAYGAAVQAAILTGDQSSTTQDLLLLDVAPLSLGIETAGGIMTKLIPRNSTIPTKKSEVFSTYADNQPGVLIQVFEGERTRTKDNNLLGKFELSGIPPAPRGVPQIEVTFDIDANGILNVSAVEKGTGKSNKITITNDKGRLSKEDIDKMVAEAEKFKAEDEQEAQRVQAKNQLESYAFTLKNSVSENNFKEKVGEEDARKLEAAAQDAINWLDASQAASTEEYKERQKELEGVANPIMSKFYGAAGGAPGAGPVPGAGAGPTGAPDNGPTVEEVD\"}, 'wesa')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
